library(dplyr)
library(data.table)
library(readr)
library(TwoSampleMR)
library(optparse)

# Load arguments ----------------------------------------------------------
# Environment: R 3.6.1

parse <- OptionParser()

option_list <- list(
   make_option('--exposure', type='character', help="GWAS sumstats for exposure", action='store'),
   make_option('--interexp', type='character', 
               help='Folder for intermediate files generated by processing exposure data', action='store')
)

args = commandArgs(trailingOnly=TRUE)

opt <- parse_args(OptionParser(option_list=option_list), args=args)
exposure.path=opt$exposure
output.path=opt$interexp

# Basic settings ----------------------------------------------------------

dir.create(file.path(output.path), showWarnings = FALSE)

# Process exposure data ---------------------------------------------------

summ.stats.fname=exposure.path
gwas_sumstats = read_tsv(summ.stats.fname) %>%
   mutate(N=2*Neff_half,Effect=log(OR)) %>%
   select(SNP,A1,A2,Effect,se=SE,p,Direction,Freq1=FRQ_A_235237,CHR,BP,N) 
traitA='MDD'

### find top SNPs

SigniSNP=gwas_sumstats %>% 
   filter(p<5e-8) %>% 
   mutate(NewAllele1=ifelse(Effect>=0,Allele1,Allele2),
          NewAllele2=ifelse(Effect>=0,Allele2,Allele1),
          NewFreq1=ifelse(Effect>=0,Freq1,1-Freq1),
          NewEffect=ifelse(Effect>=0,Effect,-1*Effect)) %>% 
   as.data.frame %>% 
   select(SNP,effect_allele=NewAllele1,other_allele=NewAllele2,
          eaf=NewFreq1,beta=NewEffect,se,pval=p,samplesize=N) %>% 
   mutate(units='unit',Phenotype=traitA)


exposure_dat = format_data(SigniSNP, type="exposure")

### clump exposure dat

exposure_dat <- clump_data(exposure_dat,clump_kb=500,clump_r2=0.001)
write_tsv(exposure_dat,paste0(output.path,'/',traitA,".exposure_dat"))

system(paste0('echo ',traitA,'   exposure prep done >> XS.log'))
      

rm(list=ls())