---
title: "PGC3 MDD Drug Enrichment"
output:
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

<style type="text/css">

h1 { /* Header 1 */
  font-size: 28px;
}
h2 { /* Header 2 */
  font-size: 22px;
}
h3 { /* Header 3 */
  font-size: 18px;
}
</style>

***

Here I test for an enrichment of genes linked to drugs. 

***

# Format DrugTargetor data

The DrugTargetor data must be converted into MAGMA .gmt format.

```{bash, eval=F, echo=T}
cd /scratch/users/k1806347/Software/mdd-meta/resources/drug_enrichment
wget https://github.com/hagax8/drugtargetor/raw/master/wholedatabase_for_targetor 
```

```{r, eval=F, echo=T}

library(data.table)
pathways<-fread('/scratch/users/k1806347/Software/mdd-meta/resources/drug_enrichment/wholedatabase_for_targetor')

length(unique(pathways$atc)) # 1763

gene_id<-fread('/users/k1806347/brc_scratch/Software/mdd-meta/docs/tables/magma.genes.out')
gene_id<-gene_id[,c('GENE','SYMBOL'),with=F]
names(gene_id)<-c('GENE','SYMBOL')

pathways<-merge(pathways,gene_id, by.x='gene',by.y='SYMBOL')

system('rm /scratch/users/k1806347/Software/mdd-meta/resources/drug_enrichment/wholedatabase_for_targetor.gmt')

drugs<-unique(pathways$atc)
for(i in 1:length(drugs)){
  drug_tmp<-pathways[pathways$atc == drugs[i],]
  drug_tmp<-drug_tmp[!duplicated(drug_tmp$gene),]
  drug_row<-t(data.frame(c(drug_tmp$atc[1], ' ', drug_tmp$GENE)))
  write.table(drug_row, '/scratch/users/k1806347/Software/mdd-meta/resources/drug_enrichment/wholedatabase_for_targetor.gmt', append=T, col.names=F, row.names=F, quote=F, sep='\t')
}

```

***

# Run MAGMA

Note. This analysis has been deprecated as we are no longer using MAGMA for gene associations.

The gene associations from MAGMA were generated by David Howard using FUMA.

```{bash, eval=F, echo=T}

/users/k1806347/brc_scratch/Software/magma_v1.09b.sh \
--gene-results /users/k1806347/brc_scratch/Software/mdd-meta/docs/tables/magma.genes.raw \
--set-annot /scratch/users/k1806347/Software/mdd-meta/resources/drug_enrichment/wholedatabase_for_targetor.gmt \
--out /scratch/users/k1806347/Software/mdd-meta/results/drug_enrichment/PGC3.MDD3.DrugGeneSet

```

```{r, eval=F, echo=T}

drug_res<-read.table('/scratch/users/k1806347/Software/mdd-meta/results/drug_enrichment/PGC3.MDD3.DrugGeneSet.gsa.out', header=T, stringsAsFactors=F)
drug_res<-drug_res[drug_res$NGENES >= 5,]
drug_res$P.BONF<-p.adjust(drug_res$P, method='bonferroni')
drug_res$P.FDR<-p.adjust(drug_res$P, method='fdr')
drug_res<-drug_res[order(drug_res$P),]

sum(drug_res$P.FDR < 0.05)

write.csv(drug_res, '/scratch/users/k1806347/Software/mdd-meta/results/drug_enrichment/PGC3.MDD3.DrugGeneSet.csv', row.names=F, quote=T)

# Test for enrichment of antidepressants are enriched
drug_res$AntiDep<-0
drug_res$AntiDep[grepl('ATC:N06A',drug_res$FULL_NAME)]<-1

drug_res$rank<-1:dim(drug_res)[1]
wilcox.test(drug_res$rank ~ drug_res$AntiDep, conf.int =T, alternative='greater')

# p-value = 0.01155
table(drug_res$AntiDep)

median(drug_res$rank[drug_res$AntiDep == 0])
median(drug_res$rank[drug_res$AntiDep == 1])

# Test for enrichment of antipsychotics are enriched
drug_res$AntiPsych<-0
drug_res$AntiPsych[grepl('ATC:N05A',drug_res$FULL_NAME)]<-1

wilcox.test(drug_res$rank ~ drug_res$AntiPsych, conf.int =T, alternative='greater')

# p-value = 0.001966

median(drug_res$rank[drug_res$AntiPsych == 0])
median(drug_res$rank[drug_res$AntiPsych == 1])

table(drug_res$AntiPsych)

```

***

# TWAS/CMAP

This approach uses the TWAS results and the CMAP/LINCS database to find drugs that induce a drug expression signiture opposite to depression. The analysis run using the clue.io query tool, which has a limit of 150 up and 150 downregulated genes.

```{r, eval=F, echo=T}
# Read in high confidence results
high_conf_res<-fread('../../results/twas/PGC3_MDD_TWAS_HighConf_results.csv')

# Remove results from splicing as direction is not clear
high_conf_res<-high_conf_res[high_conf_res$`SNP-weight Set` != 'CMC DLPFC Splicing',]

sum(high_conf_res$TWAS.Z > 0) # 24
sum(high_conf_res$TWAS.Z < 0) # 43

write.table(high_conf_res$ID[high_conf_res$TWAS.Z > 0], '../../results/twas/PGC3_MDD_TWAS_HighConf_Up_clue.txt', col.names=F, row.names=F, quote=F)
write.table(high_conf_res$ID[high_conf_res$TWAS.Z < 0], '../../results/twas/PGC3_MDD_TWAS_HighConf_Down_clue.txt', col.names=F, row.names=F, quote=F)

# This is a decent number of genes. Not all will be available in the CMAP database though.
# Also run the analysis using more relaxed criteria.
res<-read.csv('../../results/twas/conditional/PGC_MDD3_TWAS_Conditional_table_novelty.csv')

# Remove results from splicing as direction is not clear
res<-res[res$PANEL_clean != 'CMC DLPFC Splicing',]

# Remove genes that did not colocalise
res<-res[res$Colocalised == T,]

# Remove duplicates
res<-res[!duplicated(res$ID),]

sum(res$TWAS.Z > 0) # 154
sum(res$TWAS.Z < 0) # 153

# Although this is slightly too many for clue.io, some won't be in there database so this probably fine.

write.table(res$ID[res$TWAS.Z > 0], '../../results/twas/conditional/PGC_MDD3_TWAS_sig_coloc_Up_clue.txt', col.names=F, row.names=F, quote=F)
write.table(res$ID[res$TWAS.Z < 0], '../../results/twas/conditional/PGC_MDD3_TWAS_sig_coloc_Down_clue.txt', col.names=F, row.names=F, quote=F)

```

Note. when submitting the v2 run on clue.io, I got the names the wrong way around, so highconf should be sig_coloc and vice versa. I have corrected the names when downloading the files.

```{r, eval=T, echo=T}
# Read in the clue.io results
# Create function to read GCT
read_gct<-function(file){
  library(data.table)
  tmp<-fread(file, skip=2, header=T)
  tmp<-tmp[-1,]
  return(tmp)
}

highconf_gsea<-read_gct('/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment/clue/mdd_highconf_gsea_v2.gct')
highconf_query<-read_gct('/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment/clue/mdd_highconf_query_v2.gct')

sig_coloc_gsea<-read_gct('/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment/clue/mdd_sig_coloc_gsea_v2.gct')
sig_coloc_query<-read_gct('/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment/clue/mdd_sig_coloc_query_v2.gct')

# Format gsea results
highconf_gsea<-highconf_gsea[,c("src_set_id","set_type","cell_iname","pert_type","raw_cs","fdr_q_nlog10"), with=F]
highconf_gsea_sig<-highconf_gsea[highconf_gsea$fdr_q_nlog10 > 1.30103,]
highconf_gsea_sig<-highconf_gsea_sig[order(as.numeric(highconf_gsea_sig$raw_cs)),]

highconf_gsea_sig_neg<-highconf_gsea_sig[highconf_gsea_sig$raw_cs < 0,]
names(highconf_gsea_sig_neg)<-c('Gene Set', 'Set Type', 'Cell Name', 'Pertubagen Type','Connnectivity Score',"-log10(p.fdr)")

highconf_gsea_sig_pos<-highconf_gsea_sig[highconf_gsea_sig$raw_cs > 0,]
names(highconf_gsea_sig_pos)<-c('Gene Set', 'Set Type', 'Cell Name', 'Pertubagen Type','Connnectivity Score',"-log10(p.fdr)")

library(knitr)
kable(highconf_gsea_sig_neg, row.names=F , caption='Significant sets with negative connectivity score', digits = 32)
kable(highconf_gsea_sig_pos, row.names=F , caption='Significant sets with positive connectivity score', digits = 32)

sig_coloc_gsea<-sig_coloc_gsea[,c("src_set_id","set_type","cell_iname","pert_type","raw_cs","fdr_q_nlog10"), with=F]
sig_coloc_gsea_sig<-sig_coloc_gsea[sig_coloc_gsea$fdr_q_nlog10 > 1.30103,]
sig_coloc_gsea_sig<-sig_coloc_gsea_sig[order(as.numeric(sig_coloc_gsea_sig$raw_cs)),]

sig_coloc_gsea_sig_neg<-sig_coloc_gsea_sig[sig_coloc_gsea_sig$raw_cs < 0,]
names(sig_coloc_gsea_sig_neg)<-c('Gene Set', 'Set Type', 'Cell Name', 'Pertubagen Type','Connnectivity Score',"-log10(p.fdr)")

sig_coloc_gsea_sig_pos<-sig_coloc_gsea_sig[sig_coloc_gsea_sig$raw_cs > 0,]
names(sig_coloc_gsea_sig_pos)<-c('Gene Set', 'Set Type', 'Cell Name', 'Pertubagen Type','Connnectivity Score',"-log10(p.fdr)")

library(knitr)
kable(sig_coloc_gsea_sig_neg, row.names=F , caption='Significant sets with negative connectivity score', digits = 32)
kable(sig_coloc_gsea_sig_pos, row.names=F , caption='Significant sets with positive connectivity score', digits = 32)

sig_coloc_gsea[grepl('SSRI', sig_coloc_gsea$src_set_id),]
highconf_gsea[grepl('SSRI', highconf_gsea$src_set_id),]

# Format query results
highconf_query<-highconf_query[,c("pert_id","pert_iname","cell_iname","pert_type","raw_cs","fdr_q_nlog10"), with=F]
highconf_query_sig<-highconf_query[highconf_query$fdr_q_nlog10 > 1.30103,]
highconf_query_sig<-highconf_query_sig[order(as.numeric(highconf_query_sig$raw_cs)),]

sig_coloc_query<-sig_coloc_query[,c("pert_id","pert_iname","cell_iname","pert_type","raw_cs","fdr_q_nlog10"), with=F]
sig_coloc_query_sig<-sig_coloc_query[sig_coloc_query$fdr_q_nlog10 > 1.30103,]
sig_coloc_query_sig<-sig_coloc_query_sig[order(as.numeric(sig_coloc_query_sig$raw_cs)),]


highconf_query_sig_neg<-highconf_query_sig[highconf_query_sig$raw_cs < 0,]
highconf_query_sig_neg$pert_type[highconf_query_sig_neg$pert_type == 'trt_oe']<-'cDNA for overexpression of wild-type gene'
highconf_query_sig_neg$pert_type[highconf_query_sig_neg$pert_type == 'trt_sh']<-'shRNA for loss of function (LoF) of gene'
names(highconf_query_sig_neg)<-c('Pertubagen ID','Gene','Cell Name','Pertubagen Type','Connectivity Score',"-log10(p.fdr)")

highconf_query_sig_pos<-highconf_query_sig[highconf_query_sig$raw_cs > 0,]
names(highconf_query_sig_pos)<-c('Pertubagen ID','Gene','Cell Name','Pertubagen Type','Connectivity Score',"-log10(p.fdr)")

kable(highconf_query_sig_neg, row.names=F , caption='Significant pertubagens with negative connectivity score', digits = 32)
kable(highconf_query_sig_pos, row.names=F , caption='Significant pertubagens with positive connectivity score', digits = 32)

sig_coloc_query_sig_neg<-sig_coloc_query_sig[sig_coloc_query_sig$raw_cs < 0,]
sig_coloc_query_sig_neg$pert_type[sig_coloc_query_sig_neg$pert_type == 'trt_oe']<-'cDNA for overexpression of wild-type gene'
sig_coloc_query_sig_neg$pert_type[sig_coloc_query_sig_neg$pert_type == 'trt_sh']<-'shRNA for loss of function (LoF) of gene'
names(sig_coloc_query_sig_neg)<-c('Pertubagen ID','Gene','Cell Name','Pertubagen Type','Connectivity Score',"-log10(p.fdr)")

sig_coloc_query_sig_pos<-sig_coloc_query_sig[sig_coloc_query_sig$raw_cs > 0,]
names(sig_coloc_query_sig_pos)<-c('Pertubagen ID','Gene','Cell Name','Pertubagen Type','Connectivity Score',"-log10(p.fdr)")

kable(sig_coloc_query_sig_neg, row.names=F , caption='Significant pertubagens with negative connectivity score', digits = 32)
kable(sig_coloc_query_sig_pos, row.names=F , caption='Significant pertubagens with positive connectivity score', digits = 32)

# Look up antidepressant results
antidep<-c('Amitriptyline','Citalopram','Clomipramine','Dosulepin','Duloxetine','Escitalopram','Fluoxetine','Fluvoxamine','Imipramine','Maprotiline','Mianserin','Mirtazapine','Nortriptyline','Paroxetine','Reboxetine','Selegiline','Sertraline','Tranylcypromine','Trazodone','Trimipramine','Venlafaxine')

sig_coloc_query_antidep<-sig_coloc_query[grepl(paste0('^',antidep,'$', collapse='|'), x=sig_coloc_query$pert_iname, ignore.case = T),]
sig_coloc_query_antidep
highconf_query_antidep<-highconf_query[grepl(paste0('^',antidep,'$', collapse='|'), x=highconf_query$pert_iname, ignore.case = T),]
highconf_query_antidep



```

***
***

# Using GenoFuncPipe

Here we will put the PGC3MDD sumstats into the GenoFuncPipe.

***

## Format for GenoFuncPipe

```{r, eval=F, echo=T}
library(data.table)

# Read in the sumstats
ss<-fread('/users/k1806347/brc_scratch/Software/mdd-meta/results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.11.gz')

# Calculate FREQ
ss$FREQ<-((ss$FRQ_A_525197 * 525197) + (ss$FRQ_U_3362335 * 3362335))/(525197+3362335)

# Calculate Neff
ss$N<-ss$Neff_half*2
  
# Subset required columns
ss<-ss[,c('CHR','BP','SNP','A1','A2','P','OR','SE','N','FREQ','INFO'), with=F]

# Write out the results
dir.create('/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment')
fwrite(ss, '/users/k1806347/brc_scratch/Software/mdd-meta/results/drug_enrichment/daner_pgc_mdd_full_eur_hg19_v3.49.24.11_GenoFunc.gz', quote=F, sep=' ', na='NA')

```

***

# Run GenoFuncPipe

First insert GWAS information into the GenoFuncPipe gwas_list.

```{bash, eval=F, echo=T}
cd 
snakemake --profile slurm --use-conda results/PGC3MDD/reports/PGC3MDD_report.html

```


