<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<link rel="stylesheet" href="data:text/css,%0A%40font%2Dface%20%7B%0Afont%2Dfamily%3A%20octicons%2Dlink%3B%0Asrc%3A%20url%28data%3Afont%2Fwoff%3Bcharset%3Dutf%2D8%3Bbase64%2Cd09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM%2B8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB%2FaFGpk3jaTY6xa8JAGMW%2FO62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v%2Bk%2F0an2i%2BitHDw3v2%2B9%2BDBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3%2FI7AtxEJLtzzuZfI%2BVVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy%2FLt7Kc%2B0vWY%2FgAgIIEqAN9we0pwKXreiMasxvabDQMM4riO%2BqxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw%2Bymhce7vwM9jSqO8JyVd5RH9gyTt2%2BJ%2FyUmYlIR0s04n6%2B7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv%2FocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi%2BW2%2BMjCzMIDApSwvXzC97Z4Ig8N%2FBxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh%2F8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT%2BAEjAwuDFpBmA9KMDEwMCh9i%2Fv8H8sH0%2F4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9%2FlqYwOGZxeUelN2U2R6%2BcArgtCJpauW7UQBqnFkUsjAY%2FkOU1cP%2BDAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl%2BvvmM%2FbyA48e6tWrKArm4ZJlCbdsrxksL1AwWn%2FyBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO%2F%2FsdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd%2F89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF%2B9JOS0nbaaYDCQfwCJ7Au3AHj%2BLO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm%2BEBXuAbHmIMSRMs%2B4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL%2BhD7C1xoaHeLJSEao0FEW14ckxC%2BTU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13%2F%2Blm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl%2B9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O%2FAdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB%2F%2F%2FAA8AAQAAAAAAAAAAAAAAAAABAAAAAA%3D%3D%29%20format%28%27woff%27%29%3B%0A%7D%0Abody%20%7B%0A%2Dwebkit%2Dtext%2Dsize%2Dadjust%3A%20100%25%3B%0Atext%2Dsize%2Dadjust%3A%20100%25%3B%0Acolor%3A%20%23333%3B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20%22Segoe%20UI%22%2C%20Arial%2C%20freesans%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0Afont%2Dsize%3A%2016px%3B%0Aline%2Dheight%3A%201%2E6%3B%0Aword%2Dwrap%3A%20break%2Dword%3B%0A%7D%0Aa%20%7B%0Abackground%2Dcolor%3A%20transparent%3B%0A%7D%0Aa%3Aactive%2C%0Aa%3Ahover%20%7B%0Aoutline%3A%200%3B%0A%7D%0Astrong%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%202em%3B%0Amargin%3A%200%2E67em%200%3B%0A%7D%0Aimg%20%7B%0Aborder%3A%200%3B%0A%7D%0Ahr%20%7B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Aheight%3A%200%3B%0A%7D%0Apre%20%7B%0Aoverflow%3A%20auto%3B%0A%7D%0Acode%2C%0Akbd%2C%0Apre%20%7B%0Afont%2Dfamily%3A%20monospace%2C%20monospace%3B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ainput%20%7B%0Acolor%3A%20inherit%3B%0Afont%3A%20inherit%3B%0Amargin%3A%200%3B%0A%7D%0Ahtml%20input%5Bdisabled%5D%20%7B%0Acursor%3A%20default%3B%0A%7D%0Ainput%20%7B%0Aline%2Dheight%3A%20normal%3B%0A%7D%0Ainput%5Btype%3D%22checkbox%22%5D%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0Apadding%3A%200%3B%0A%7D%0Atable%20%7B%0Aborder%2Dcollapse%3A%20collapse%3B%0Aborder%2Dspacing%3A%200%3B%0A%7D%0Atd%2C%0Ath%20%7B%0Apadding%3A%200%3B%0A%7D%0A%2A%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0A%7D%0Ainput%20%7B%0Afont%3A%2013px%20%2F%201%2E4%20Helvetica%2C%20arial%2C%20nimbussansl%2C%20liberationsans%2C%20freesans%2C%20clean%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%234078c0%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%2C%0Aa%3Aactive%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200%3B%0Amargin%3A%2015px%200%3B%0Aoverflow%3A%20hidden%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23ddd%3B%0A%7D%0Ahr%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ahr%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%2015px%3B%0Amargin%2Dbottom%3A%2015px%3B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%2030px%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%2021px%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%2016px%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%2011px%3B%0A%7D%0Ablockquote%20%7B%0Amargin%3A%200%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Aol%20ol%2C%0Aul%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Droman%3B%0A%7D%0Aul%20ul%20ol%2C%0Aul%20ol%20ol%2C%0Aol%20ul%20ol%2C%0Aol%20ol%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Dalpha%3B%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0Afont%3A%2012px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0A%7D%0A%2Eselect%3A%3A%2Dms%2Dexpand%20%7B%0Aopacity%3A%200%3B%0A%7D%0A%2Eocticon%20%7B%0Afont%3A%20normal%20normal%20normal%2016px%2F1%20octicons%2Dlink%3B%0Adisplay%3A%20inline%2Dblock%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Drendering%3A%20auto%3B%0A%2Dwebkit%2Dfont%2Dsmoothing%3A%20antialiased%3B%0A%2Dmoz%2Dosx%2Dfont%2Dsmoothing%3A%20grayscale%3B%0A%2Dwebkit%2Duser%2Dselect%3A%20none%3B%0A%2Dmoz%2Duser%2Dselect%3A%20none%3B%0A%2Dms%2Duser%2Dselect%3A%20none%3B%0Auser%2Dselect%3A%20none%3B%0A%7D%0A%2Eocticon%2Dlink%3Abefore%20%7B%0Acontent%3A%20%27%5Cf05c%27%3B%0A%7D%0A%2Emarkdown%2Dbody%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%20%21important%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%20%21important%3B%0A%7D%0Aa%3Anot%28%5Bhref%5D%29%20%7B%0Acolor%3A%20inherit%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0A%2Eanchor%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%2Dright%3A%202px%3B%0Amargin%2Dleft%3A%20%2D18px%3B%0A%7D%0A%2Eanchor%3Afocus%20%7B%0Aoutline%3A%20none%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%201em%3B%0Amargin%2Dbottom%3A%2016px%3B%0Afont%2Dweight%3A%20bold%3B%0Aline%2Dheight%3A%201%2E4%3B%0A%7D%0Ah1%20%2Eocticon%2Dlink%2C%0Ah2%20%2Eocticon%2Dlink%2C%0Ah3%20%2Eocticon%2Dlink%2C%0Ah4%20%2Eocticon%2Dlink%2C%0Ah5%20%2Eocticon%2Dlink%2C%0Ah6%20%2Eocticon%2Dlink%20%7B%0Acolor%3A%20%23000%3B%0Avertical%2Dalign%3A%20middle%3B%0Avisibility%3A%20hidden%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%2C%0Ah2%3Ahover%20%2Eanchor%2C%0Ah3%3Ahover%20%2Eanchor%2C%0Ah4%3Ahover%20%2Eanchor%2C%0Ah5%3Ahover%20%2Eanchor%2C%0Ah6%3Ahover%20%2Eanchor%20%7B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah2%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah3%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah4%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah5%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah6%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%20%7B%0Avisibility%3A%20visible%3B%0A%7D%0Ah1%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%202%2E25em%3B%0Aline%2Dheight%3A%201%2E2%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah1%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah2%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%201%2E75em%3B%0Aline%2Dheight%3A%201%2E225%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah2%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%201%2E5em%3B%0Aline%2Dheight%3A%201%2E43%3B%0A%7D%0Ah3%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%201%2E25em%3B%0A%7D%0Ah4%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ah5%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%201em%3B%0Acolor%3A%20%23777%3B%0A%7D%0Ah6%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ap%2C%0Ablockquote%2C%0Aul%2C%0Aol%2C%0Adl%2C%0Atable%2C%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%204px%3B%0Apadding%3A%200%3B%0Amargin%3A%2016px%200%3B%0Abackground%2Dcolor%3A%20%23e7e7e7%3B%0Aborder%3A%200%20none%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%2Dleft%3A%202em%3B%0A%7D%0Aul%20ul%2C%0Aul%20ol%2C%0Aol%20ol%2C%0Aol%20ul%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Ali%3Ep%20%7B%0Amargin%2Dtop%3A%2016px%3B%0A%7D%0Adl%20%7B%0Apadding%3A%200%3B%0A%7D%0Adl%20dt%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%2016px%3B%0Afont%2Dsize%3A%201em%3B%0Afont%2Dstyle%3A%20italic%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Adl%20dd%20%7B%0Apadding%3A%200%2016px%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ablockquote%20%7B%0Apadding%3A%200%2015px%3B%0Acolor%3A%20%23777%3B%0Aborder%2Dleft%3A%204px%20solid%20%23ddd%3B%0A%7D%0Ablockquote%3E%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Ablockquote%3E%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Atable%20%7B%0Adisplay%3A%20block%3B%0Awidth%3A%20100%25%3B%0Aoverflow%3A%20auto%3B%0Aword%2Dbreak%3A%20normal%3B%0Aword%2Dbreak%3A%20keep%2Dall%3B%0A%7D%0Atable%20th%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Atable%20th%2C%0Atable%20td%20%7B%0Apadding%3A%206px%2013px%3B%0Aborder%3A%201px%20solid%20%23ddd%3B%0A%7D%0Atable%20tr%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Aborder%2Dtop%3A%201px%20solid%20%23ccc%3B%0A%7D%0Atable%20tr%3Anth%2Dchild%282n%29%20%7B%0Abackground%2Dcolor%3A%20%23f8f8f8%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Abackground%2Dcolor%3A%20%23fff%3B%0A%7D%0Acode%20%7B%0Apadding%3A%200%3B%0Apadding%2Dtop%3A%200%2E2em%3B%0Apadding%2Dbottom%3A%200%2E2em%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%2085%25%3B%0Abackground%2Dcolor%3A%20rgba%280%2C0%2C0%2C0%2E04%29%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0Acode%3Abefore%2C%0Acode%3Aafter%20%7B%0Aletter%2Dspacing%3A%20%2D0%2E2em%3B%0Acontent%3A%20%22%5C00a0%22%3B%0A%7D%0Apre%3Ecode%20%7B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%20100%25%3B%0Aword%2Dbreak%3A%20normal%3B%0Awhite%2Dspace%3A%20pre%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0A%2Ehighlight%20%7B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0A%2Ehighlight%20pre%2C%0Apre%20%7B%0Apadding%3A%2016px%3B%0Aoverflow%3A%20auto%3B%0Afont%2Dsize%3A%2085%25%3B%0Aline%2Dheight%3A%201%2E45%3B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0A%2Ehighlight%20pre%20%7B%0Amargin%2Dbottom%3A%200%3B%0Aword%2Dbreak%3A%20normal%3B%0A%7D%0Apre%20%7B%0Aword%2Dwrap%3A%20normal%3B%0A%7D%0Apre%20code%20%7B%0Adisplay%3A%20inline%3B%0Amax%2Dwidth%3A%20initial%3B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Aoverflow%3A%20initial%3B%0Aline%2Dheight%3A%20inherit%3B%0Aword%2Dwrap%3A%20normal%3B%0Abackground%2Dcolor%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0Apre%20code%3Abefore%2C%0Apre%20code%3Aafter%20%7B%0Acontent%3A%20normal%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%2Dsize%3A%2011px%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Epl%2Dc%20%7B%0Acolor%3A%20%23969896%3B%0A%7D%0A%2Epl%2Dc1%2C%0A%2Epl%2Ds%20%2Epl%2Dv%20%7B%0Acolor%3A%20%230086b3%3B%0A%7D%0A%2Epl%2De%2C%0A%2Epl%2Den%20%7B%0Acolor%3A%20%23795da3%3B%0A%7D%0A%2Epl%2Ds%20%2Epl%2Ds1%2C%0A%2Epl%2Dsmi%20%7B%0Acolor%3A%20%23333%3B%0A%7D%0A%2Epl%2Dent%20%7B%0Acolor%3A%20%2363a35c%3B%0A%7D%0A%2Epl%2Dk%20%7B%0Acolor%3A%20%23a71d5d%3B%0A%7D%0A%2Epl%2Dpds%2C%0A%2Epl%2Ds%2C%0A%2Epl%2Ds%20%2Epl%2Dpse%20%2Epl%2Ds1%2C%0A%2Epl%2Dsr%2C%0A%2Epl%2Dsr%20%2Epl%2Dcce%2C%0A%2Epl%2Dsr%20%2Epl%2Dsra%2C%0A%2Epl%2Dsr%20%2Epl%2Dsre%20%7B%0Acolor%3A%20%23183691%3B%0A%7D%0A%2Epl%2Dv%20%7B%0Acolor%3A%20%23ed6a43%3B%0A%7D%0A%2Epl%2Did%20%7B%0Acolor%3A%20%23b52a1d%3B%0A%7D%0A%2Epl%2Dii%20%7B%0Abackground%2Dcolor%3A%20%23b52a1d%3B%0Acolor%3A%20%23f8f8f8%3B%0A%7D%0A%2Epl%2Dsr%20%2Epl%2Dcce%20%7B%0Acolor%3A%20%2363a35c%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dml%20%7B%0Acolor%3A%20%23693a17%3B%0A%7D%0A%2Epl%2Dmh%2C%0A%2Epl%2Dmh%20%2Epl%2Den%2C%0A%2Epl%2Dms%20%7B%0Acolor%3A%20%231d3e81%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmq%20%7B%0Acolor%3A%20%23008080%3B%0A%7D%0A%2Epl%2Dmi%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dstyle%3A%20italic%3B%0A%7D%0A%2Epl%2Dmb%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmd%20%7B%0Abackground%2Dcolor%3A%20%23ffecec%3B%0Acolor%3A%20%23bd2c00%3B%0A%7D%0A%2Epl%2Dmi1%20%7B%0Abackground%2Dcolor%3A%20%23eaffea%3B%0Acolor%3A%20%2355a532%3B%0A%7D%0A%2Epl%2Dmdr%20%7B%0Acolor%3A%20%23795da3%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmo%20%7B%0Acolor%3A%20%231d3e81%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%3A%2011px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20%7B%0Alist%2Dstyle%2Dtype%3A%20none%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%2B%2Etask%2Dlist%2Ditem%20%7B%0Amargin%2Dtop%3A%203px%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20input%20%7B%0Amargin%3A%200%200%2E35em%200%2E25em%20%2D1%2E6em%3B%0Avertical%2Dalign%3A%20middle%3B%0A%7D%0A%3Achecked%2B%2Eradio%2Dlabel%20%7B%0Az%2Dindex%3A%201%3B%0Aposition%3A%20relative%3B%0Aborder%2Dcolor%3A%20%234078c0%3B%0A%7D%0A%2EsourceLine%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0A%7D%0Acode%20%2Ekw%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Edt%20%7B%20color%3A%20%23ed6a43%3B%20%7D%0Acode%20%2Edv%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Ebn%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Efl%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Ech%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Est%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Eco%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%2Eot%20%7B%20color%3A%20%230086b3%3B%20%7D%0Acode%20%2Eal%20%7B%20color%3A%20%23a61717%3B%20%7D%0Acode%20%2Efu%20%7B%20color%3A%20%2363a35c%3B%20%7D%0Acode%20%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%0Acode%20%2Ewa%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Ecn%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Esc%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Evs%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Ess%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Eim%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eva%20%7Bcolor%3A%20%23008080%3B%20%7D%0Acode%20%2Ecf%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eop%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Ebu%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eex%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Epp%20%7B%20color%3A%20%23999999%3B%20%7D%0Acode%20%2Eat%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Edo%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%2Ean%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Ecv%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Ein%20%7B%20color%3A%20%23008080%3B%20%7D%0A">
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="conditional-and-joint-analysis">Conditional and Joint Analysis</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span></code></pre></div>
<h1 id="methods">Methods</h1>
<p>We ran a <a href="https://www.nature.com/articles/ng.2213">conditional and joint analysis</a> using <a href="https://cnsgenomics.com/software/gcta/#COJO">GCTA</a> to refine the list of independent loci.</p>
<ul>
<li>Final meta-analysed SNPs from the Ricopili pipeline were used.</li>
<li>Ricopili was used for initial clumping with index SNPs identified with <em>p</em>&lt; 10^{-4} and <em>r</em><sup>2</sup>&lt; 0.1 within 3000kb windows. The extended MHC region was clumped as a single region.</li>
<li>Sumstats were filtered for MAF &gt;= 0.01, INFO &gt; 0.6, and Neff &gt;= 80% of max.</li>
<li>Regions with a genome-wide significant SNP (p &lt; 5-e8) were identified from the clumped results. Regions within 50kb of each other were merged.</li>
<li>SNPs from these regions were extracted and filtered to unrelated of self- and genotype-identified European ancestry participants from UK Biobank.</li>
<li>A conditional analysis was performed on each region using the filtered sumstats superimposed on the UK Biobank LD structure.</li>
<li>Singleton regions (with only one genome-wide significant variant) were removed.</li>
</ul>
<h2 id="previous-sumstats">Previous sumstats</h2>
<p>Variants from <a href="https://www.nature.com/articles/s41593-018-%200326-7">Howard et al 2019</a> (remove first and last rows with table captions) and <a href="https://doi.org/10.1038/s41593-021-00860-2">Levey et al 2021</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>howard <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>howard, <span class="at">skip=</span><span class="dv">2</span>, <span class="at">n_max=</span><span class="dv">102</span>)</span></code></pre></div>
<pre><code>## New names:
## * `Odds Ratio` -&gt; `Odds Ratio...8`
## * `Lower 95% Confidence Interval` -&gt; `Lower 95% Confidence Interval...9`
## * `Upper 95% Confidence Interval` -&gt; `Upper 95% Confidence Interval...10`
## * `Log(Odds Ratio)` -&gt; `Log(Odds Ratio)...11`
## * `Standard error of the Log(Odds Ratio)` -&gt; `Standard error of the Log(Odds Ratio)...12`
## * ...</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>levey <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>levey, <span class="at">col_types=</span><span class="fu">cols</span>(<span class="at">CHR.BP=</span><span class="fu">col_character</span>()))</span></code></pre></div>
<h1 id="results">Results</h1>
<h2 id="cojo-snp-and-region-counts">COJO SNP and region counts</h2>
<p>List of clumped and COJO SNPs and regions</p>
<ul>
<li>Sumstats: results/cojo/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.qc.gz</li>
<li>Clump file: results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc</li>
<li>COJO regions: 577</li>
<li>Clumped SNPs: 817</li>
<li>COJO Selected SNPs: 617</li>
<li>Singleton regions: 65</li>
<li>COJO+Clump SNPs: 617</li>
<li>COJO Final SNPs: 552</li>
<li>COJO Final SNPs p &lt;= 5e-8, pJ &lt;= 5e-8: 543</li>
<li>COJO Final SNPs p &gt; 5e-8, pJ &lt;= 5e-8: 9</li>
</ul>
<p>Load list of COJO SNPs</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>cojo <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>cojo)</span></code></pre></div>
<pre><code>## 
## ── Column specification ───────────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   SNP = col_character(),
##   A1 = col_character(),
##   A2 = col_character(),
##   Direction = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<h2 id="clumped-results">Clumped results</h2>
<p>Clumped results from Ricopili</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rp <span class="ot">&lt;-</span> <span class="fu">read_table2</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>rp_clump) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(P <span class="sc">&lt;=</span> <span class="fl">5e-8</span>)</span></code></pre></div>
<pre><code>## 
## ── Column specification ───────────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   SNP = col_character(),
##   A1A2 = col_character(),
##   `(Nca,Nco,Neff)Dir` = col_character(),
##   `LD-friends(0.1).p0.001` = col_character(),
##   `LD-friends(0.6).p0.001` = col_character(),
##   gwas_catalog_span.6 = col_character(),
##   `genes.6.50kb(dist2index)` = col_character(),
##   N.genes.6.50kb = col_character()
## )
## ℹ Use `spec()` for the full column specifications.

## Warning: 94 parsing failures.
##  row col expected actual                                                                                           file
## 1420 ngt a double      - &#39;results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc&#39;
## 1850 ngt a double      - &#39;results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc&#39;
## 1919 ngt a double      - &#39;results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc&#39;
## 2244 ngt a double      - &#39;results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc&#39;
## 2345 ngt a double      - &#39;results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.09.gz.p4.clump.areator.sorted.1mhc&#39;
## .... ... ........ ...... ..............................................................................................
## See problems(...) for more details.</code></pre>
<h2 id="genomic-ranges">Genomic ranges</h2>
<p>Construct genomic range objects so that SNPs and regions can be intersected and compared.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>cojo_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(cojo, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>range.left, <span class="at">end=</span>range.right), <span class="at">SNP=</span>SNP))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>rp_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(rp, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>range.left, <span class="at">end=</span>range.right), <span class="at">SNP=</span>SNP))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>howard_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(howard, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>Chromosome, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span><span class="st">`</span><span class="at">Postion (bp)</span><span class="st">`</span>, <span class="at">width=</span><span class="dv">1</span>)))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>levey_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(levey, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>BP, <span class="at">width=</span><span class="dv">1</span>)))</span></code></pre></div>
<h2 id="gwas-catalog">GWAS catalog</h2>
<p>Tally entries in the GWAS catalog for each region</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>parse_catalog_entry <span class="ot">&lt;-</span> <span class="cf">function</span>(catalog_entry) {</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># extract LD and RSID from first element</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    ld_snp_match <span class="ot">&lt;-</span> <span class="fu">str_match</span>(catalog_entry[<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)(rs[[:digit:]]+)&quot;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    ld <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(ld_snp_match[,<span class="dv">2</span>])</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    snp <span class="ot">&lt;-</span> ld_snp_match[,<span class="dv">3</span>]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># parse rest of elements into phenotype(PubMed ID)(P-value)</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    phenotype_matches <span class="ot">&lt;-</span> <span class="fu">str_match</span>(catalog_entry[<span class="sc">-</span><span class="dv">1</span>], <span class="st">&quot;(.+)</span><span class="sc">\\</span><span class="st">(([[:digit:]]+)</span><span class="sc">\\</span><span class="st">)</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)&quot;</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    phenotypes <span class="ot">&lt;-</span> phenotype_matches[,<span class="dv">2</span>]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    pubmeds <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenotype_matches[,<span class="dv">3</span>])</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    p_values <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenotype_matches[,<span class="dv">4</span>])</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">data.frame</span>(ld, <span class="at">catalogSNP=</span>snp, <span class="at">phenotype=</span>phenotypes, <span class="at">pubmed_id=</span>pubmeds, <span class="at">P=</span>p_values))</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>rp_gwas_catalog_entries <span class="ot">&lt;-</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>plyr<span class="sc">::</span><span class="fu">ddply</span>(<span class="fu">filter</span>(rp, gwas_catalog_span<span class="fl">.6</span> <span class="sc">!=</span> <span class="st">&#39;-&#39;</span>), <span class="sc">~</span>SNP, <span class="cf">function</span>(rp_entry) {</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get GWAS catalog cell</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    gwas_catalog <span class="ot">&lt;-</span> rp_entry<span class="sc">$</span>gwas_catalog_span<span class="fl">.6</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># split out into catalog entries (separated by /, removing first empty element)</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    catalog_entries <span class="ot">&lt;-</span> <span class="fu">str_split</span>(gwas_catalog, <span class="st">&quot;/&quot;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    catalog_entries_complete <span class="ot">&lt;-</span> catalog_entries[<span class="fu">which</span>(catalog_entries <span class="sc">!=</span> <span class="st">&quot;&quot;</span>)]</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># split SNP entries by &quot;;&quot;</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    catalog_entries_list <span class="ot">&lt;-</span> <span class="fu">str_split</span>(catalog_entries_complete, <span class="st">&quot;;&quot;</span>)</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># parse each entry</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    catalog_entries_df <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">ldply</span>(catalog_entries_list, parse_catalog_entry)</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(catalog_entries_df)</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<h1 id="parse-gene-list">Parse gene list</h1>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="ot">&lt;-</span> </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>plyr<span class="sc">::</span><span class="fu">ddply</span>(<span class="fu">filter</span>(rp, <span class="st">`</span><span class="at">genes.6.50kb(dist2index)</span><span class="st">`</span> <span class="sc">!=</span> <span class="st">&#39;-&#39;</span>), <span class="sc">~</span>SNP, <span class="cf">function</span>(rp_entry) {</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    genes_dist <span class="ot">&lt;-</span> rp_entry<span class="sc">$</span><span class="st">`</span><span class="at">genes.6.50kb(dist2index)</span><span class="st">`</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    genes_dist_list <span class="ot">&lt;-</span> <span class="fu">str_split</span>(genes_dist, <span class="st">&#39;,&#39;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    genes_dist_match <span class="ot">&lt;-</span> <span class="fu">str_match</span>(genes_dist_list, <span class="st">&quot;(.+)</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)&quot;</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="at">gene=</span>genes_dist_match[,<span class="dv">2</span>], <span class="at">dist2index=</span><span class="fu">as.numeric</span>(genes_dist_match[,<span class="dv">3</span>])))</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<h2 id="comparison-to-previous-findings">Comparison to previous findings</h2>
<p>Find which COJO regions overlap with Howard</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>cojo_howard_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(cojo_gr, howard_gr)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>cojo_howard_overlaps</span></code></pre></div>
<pre><code>## Hits object with 115 hits and 0 metadata columns:
##         queryHits subjectHits
##         &lt;integer&gt;   &lt;integer&gt;
##     [1]         1           1
##     [2]         5           2
##     [3]         6           3
##     [4]         8           4
##     [5]         9           4
##     ...       ...         ...
##   [111]       521          97
##   [112]       525          98
##   [113]       532         100
##   [114]       534         101
##   [115]       548         102
##   -------
##   queryLength: 552 / subjectLength: 102</code></pre>
<p>Count number of regions in Howard that overlap:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>howard <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_howard_overlaps<span class="sc">@</span>to)) <span class="sc">%&gt;%</span> <span class="fu">count</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 × 1
##       n
##   &lt;int&gt;
## 1    88</code></pre>
<p>Find which COJO regions overlap with Levey</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>cojo_levey_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(cojo_gr, levey_gr)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>cojo_levey_overlaps</span></code></pre></div>
<pre><code>## Hits object with 260 hits and 0 metadata columns:
##         queryHits subjectHits
##         &lt;integer&gt;   &lt;integer&gt;
##     [1]         1         110
##     [2]         5           2
##     [3]         6          65
##     [4]         8          77
##     [5]         8          26
##     ...       ...         ...
##   [256]       537         154
##   [257]       538         154
##   [258]       544         120
##   [259]       548         153
##   [260]       549         211
##   -------
##   queryLength: 552 / subjectLength: 223</code></pre>
<p>Count number of regions in Levey that overlap:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>levey <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_levey_overlaps<span class="sc">@</span>to)) <span class="sc">%&gt;%</span> <span class="fu">count</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 × 1
##       n
##   &lt;int&gt;
## 1   191</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>cojo_known <span class="ot">&lt;-</span> cojo <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_levey_overlaps<span class="sc">@</span>from))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>catalog_known <span class="ot">&lt;-</span> rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_known<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>catalog_known</span></code></pre></div>
<pre><code>## # A tibble: 166 × 2
##    phenotype                   n
##    &lt;chr&gt;                   &lt;int&gt;
##  1 Serum_metabolit...        186
##  2 Schizophrenia              45
##  3 Intelligence_(MTAG)        42
##  4 Trans_fatty_acid_levels    29
##  5 Waist_circumfer...         29
##  6 Depression_(broad)         28
##  7 Glycerophosphol...         27
##  8 Neuroticism                26
##  9 Autism_spectrum...         24
## 10 Depression                 22
## # … with 156 more rows</code></pre>
<p>Newly discovered regions</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>cojo_new <span class="ot">&lt;-</span> cojo <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(<span class="sc">-</span>cojo_levey_overlaps<span class="sc">@</span>from)) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(P)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>cojo_new <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(region, snp_idx, CHR, SNP, BP, P, pJ) <span class="sc">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">group_by</span>(region)</span></code></pre></div>
<pre><code>## # A tibble: 360 × 7
## # Groups:   region [342]
##    region snp_idx   CHR SNP               BP        P       pJ
##     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1    174       1     5 rs1993739  153215007 5.71e-20 5.72e-20
##  2    487       1    20 rs17805843  51205902 9.34e-20 9.36e-20
##  3     49       2     2 rs359247    60477052 2.67e-18 6.60e-17
##  4    166       1     5 rs6863440  120084621 3.25e-18 3.26e-18
##  5    211       1     7 rs10499337   3521803 1.19e-17 1.48e-16
##  6    362       1    12 rs2363585   60791165 1.69e-16 1.69e-16
##  7    310       1    10 rs12778915  77617557 2.35e-16 2.35e-16
##  8    374       1    12 rs7962128  121907336 3.68e-16 3.68e-16
##  9     76       1     2 rs13418032 198413692 9.89e-16 9.90e-16
## 10     81       1     2 rs72931605 212693775 1.29e-15 1.21e- 9
## # … with 350 more rows</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_new<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n)) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span>phenotype <span class="sc">%in%</span> catalog_known<span class="sc">$</span>phenotype)</span></code></pre></div>
<pre><code>## # A tibble: 61 × 2
##    phenotype                    n
##    &lt;chr&gt;                    &lt;int&gt;
##  1 Body_mass_index...           9
##  2 Morning_vs._eve...           5
##  3 Obesity                      5
##  4 Hip_circumference            4
##  5 Diastolic_blood_pressure     3
##  6 Monocyte_count               3
##  7 Pulse_pressure               3
##  8 Ulcerative_colitis           3
##  9 Waist_circumference          3
## 10 Adiponectin_levels           2
## # … with 51 more rows</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_new<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(SNP) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(dist2index <span class="sc">==</span> <span class="fu">min</span>(dist2index)) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="fu">select</span>(gene) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>()</span></code></pre></div>
<pre><code>## # A tibble: 374 × 1
##    gene    
##    &lt;chr&gt;   
##  1 NISCH   
##  2 STAB1   
##  3 NT5DC2  
##  4 SMIM4   
##  5 PBRM1   
##  6 NTRK3   
##  7 GTF2IRD1
##  8 NXPH1   
##  9 MAGI2   
## 10 ETV6    
## # … with 364 more rows</code></pre>
<h2 id="clumped-results-1">Clumped results</h2>
<p>Find regions overlapping between clumped and COJO results</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cojo_clumped_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(cojo_gr, rp_gr)</span></code></pre></div>
<p>List COJO results where the selected SNP is not in the clumped results</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>cojo_newly_selected <span class="ot">&lt;-</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>cojo <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_clumped_overlaps<span class="sc">@</span>from)) <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="sc">!</span>SNP <span class="sc">%in%</span> rp<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(region, snp_idx, CHR, SNP, P, pJ) </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>cojo_newly_selected</span></code></pre></div>
<pre><code>## # A tibble: 58 × 6
##    region snp_idx   CHR SNP                P       pJ
##     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt;    &lt;dbl&gt;
##  1      1       1     1 rs301806    1.87e-16 1.87e-16
##  2     10       1     1 rs437021    5.70e-11 5.71e-11
##  3     13       1     1 rs3101341   5.22e-27 2.29e-11
##  4     13       2     1 rs2797104   5.84e-32 4.16e-10
##  5     47       1     2 rs858938    3.87e- 8 2.65e-10
##  6     48       2     2 rs56873970  2.76e- 3 1.82e- 8
##  7     49       1     2 rs7576017   1.18e-13 2.96e-12
##  8     50       1     2 rs149044563 7.75e-13 7.76e-13
##  9     53       1     2 rs73949838  5.43e- 9 5.43e- 9
## 10     55       1     2 rs3860446   1.10e-10 3.29e-11
## # … with 48 more rows</code></pre>
<p>Newly selected SNPs that were not GWsig in the clumped results</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>cojo_newly_selected <span class="sc">%&gt;%</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(P <span class="sc">&gt;</span> <span class="fl">5e-8</span>)</span></code></pre></div>
<pre><code>## # A tibble: 9 × 6
##   region snp_idx   CHR SNP                   P       pJ
##    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;             &lt;dbl&gt;    &lt;dbl&gt;
## 1     48       2     2 rs56873970 0.00276      1.82e- 8
## 2     60       1     2 rs2381462  0.00000896   9.62e-11
## 3     69       2     2 rs1371187  0.0000000835 4.50e-11
## 4    118       1     3 rs13073224 0.0000000573 2.02e- 8
## 5    186       2     6 rs2747467  0.144        1.63e-12
## 6    248       1     8 rs10503484 0.000000367  7.61e- 9
## 7    341       4    11 rs73004019 0.000154     5.15e-10
## 8    386       2    13 rs2329076  0.000000154  3.24e- 8
## 9    412       2    15 rs4774501  0.00000214   6.16e-11</code></pre>
<p>List clumped SNPs in retained regions that were not selected by COJO</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>rp <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_clumped_overlaps<span class="sc">@</span>to)) <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="sc">!</span>SNP <span class="sc">%in%</span> cojo<span class="sc">$</span>SNP)</span></code></pre></div>
<pre><code>## # A tibble: 246 × 23
##    SNP          CHR       BP        P    OR     SE A1A2  FRQ_A_524857 FRQ_U_3059006
##    &lt;chr&gt;      &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;         &lt;dbl&gt;
##  1 rs301817       1  8503379 3.23e-17 1.02  0.0028 C/A         0.419         0.426 
##  2 rs75986133     1 52846058 1.99e- 8 0.967 0.0061 A/G         0.044         0.0458
##  3 rs446952       1 61738636 1.78e-11 1.02  0.0027 T/C         0.465         0.458 
##  4 rs2568957      1 72764430 1.58e-32 0.963 0.0031 A/G         0.176         0.172 
##  5 rs12127789     1 72740073 4.91e-23 1.04  0.0038 T/G         0.119         0.119 
##  6 rs12748090     1 73005277 1.47e-14 1.02  0.003  A/T         0.279         0.277 
##  7 rs75805282     1 73857518 6.32e-13 0.950 0.0071 T/C         0.0301        0.0304
##  8 rs61771936     1 73266056 3.66e-12 0.977 0.0034 A/G         0.156         0.158 
##  9 rs12128239     1 73059446 4.20e-12 0.977 0.0034 G/A         0.163         0.164 
## 10 rs10736420     1 74000649 1.18e-11 1.03  0.0036 C/T         0.12          0.118 
## # … with 236 more rows, and 14 more variables: INFO &lt;dbl&gt;,
## #   (Nca,Nco,Neff)Dir &lt;chr&gt;, ngt &lt;dbl&gt;, LD-friends(0.1).p0.001 &lt;chr&gt;,
## #   range.left &lt;dbl&gt;, range.right &lt;dbl&gt;, span(kb) &lt;dbl&gt;,
## #   LD-friends(0.6).p0.001 &lt;chr&gt;, range.left.6 &lt;dbl&gt;, range.right.6 &lt;dbl&gt;,
## #   span.6(kb) &lt;dbl&gt;, gwas_catalog_span.6 &lt;chr&gt;,
## #   genes.6.50kb(dist2index) &lt;chr&gt;, N.genes.6.50kb &lt;chr&gt;</code></pre>
<p>Line up non-selected SNPs with selected SNPs in the region</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bind_cols</span>(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(<span class="fu">slice</span>(cojo, cojo_clumped_overlaps<span class="sc">@</span>from), region, snp_idx, <span class="at">SNP.cojo=</span>SNP, <span class="at">BP.cojo=</span>BP, <span class="at">P.cojo=</span>P, <span class="at">PJ.cojo=</span>pJ),</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(<span class="fu">slice</span>(rp, cojo_clumped_overlaps<span class="sc">@</span>to), <span class="at">SNP.rp=</span>SNP, <span class="at">BP.rp=</span>BP, <span class="at">P.rp=</span>P)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(<span class="sc">!</span>SNP.rp <span class="sc">%in%</span> cojo<span class="sc">$</span>SNP)</span></code></pre></div>
<pre><code>## # A tibble: 400 × 9
##    region snp_idx SNP.cojo   BP.cojo   P.cojo  PJ.cojo SNP.rp      BP.rp     P.rp
##     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;
##  1      1       1 rs301806   8482078 1.87e-16 1.87e-16 rs301817   8.50e6 3.23e-17
##  2      9       1 rs7413471 52339759 2.96e-15 2.96e-15 rs75986133 5.28e7 1.99e- 8
##  3     10       1 rs437021  61738270 5.70e-11 5.71e-11 rs446952   6.17e7 1.78e-11
##  4     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs2568957  7.28e7 1.58e-32
##  5     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs12127789 7.27e7 4.91e-23
##  6     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs12748090 7.30e7 1.47e-14
##  7     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs75805282 7.39e7 6.32e-13
##  8     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs61771936 7.33e7 3.66e-12
##  9     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs12128239 7.31e7 4.20e-12
## 10     13       1 rs3101341 72747844 5.22e-27 2.29e-11 rs10736420 7.40e7 1.18e-11
## # … with 390 more rows</code></pre>

</body>
</html>
