<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<link rel="stylesheet" href="data:text/css,%0A%40font%2Dface%20%7B%0Afont%2Dfamily%3A%20octicons%2Dlink%3B%0Asrc%3A%20url%28data%3Afont%2Fwoff%3Bcharset%3Dutf%2D8%3Bbase64%2Cd09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM%2B8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB%2FaFGpk3jaTY6xa8JAGMW%2FO62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v%2Bk%2F0an2i%2BitHDw3v2%2B9%2BDBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3%2FI7AtxEJLtzzuZfI%2BVVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy%2FLt7Kc%2B0vWY%2FgAgIIEqAN9we0pwKXreiMasxvabDQMM4riO%2BqxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw%2Bymhce7vwM9jSqO8JyVd5RH9gyTt2%2BJ%2FyUmYlIR0s04n6%2B7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv%2FocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi%2BW2%2BMjCzMIDApSwvXzC97Z4Ig8N%2FBxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh%2F8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT%2BAEjAwuDFpBmA9KMDEwMCh9i%2Fv8H8sH0%2F4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9%2FlqYwOGZxeUelN2U2R6%2BcArgtCJpauW7UQBqnFkUsjAY%2FkOU1cP%2BDAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl%2BvvmM%2FbyA48e6tWrKArm4ZJlCbdsrxksL1AwWn%2FyBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO%2F%2FsdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd%2F89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF%2B9JOS0nbaaYDCQfwCJ7Au3AHj%2BLO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm%2BEBXuAbHmIMSRMs%2B4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL%2BhD7C1xoaHeLJSEao0FEW14ckxC%2BTU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13%2F%2Blm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl%2B9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O%2FAdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB%2F%2F%2FAA8AAQAAAAAAAAAAAAAAAAABAAAAAA%3D%3D%29%20format%28%27woff%27%29%3B%0A%7D%0Abody%20%7B%0A%2Dwebkit%2Dtext%2Dsize%2Dadjust%3A%20100%25%3B%0Atext%2Dsize%2Dadjust%3A%20100%25%3B%0Acolor%3A%20%23333%3B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20%22Segoe%20UI%22%2C%20Arial%2C%20freesans%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0Afont%2Dsize%3A%2016px%3B%0Aline%2Dheight%3A%201%2E6%3B%0Aword%2Dwrap%3A%20break%2Dword%3B%0A%7D%0Aa%20%7B%0Abackground%2Dcolor%3A%20transparent%3B%0A%7D%0Aa%3Aactive%2C%0Aa%3Ahover%20%7B%0Aoutline%3A%200%3B%0A%7D%0Astrong%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%202em%3B%0Amargin%3A%200%2E67em%200%3B%0A%7D%0Aimg%20%7B%0Aborder%3A%200%3B%0A%7D%0Ahr%20%7B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Aheight%3A%200%3B%0A%7D%0Apre%20%7B%0Aoverflow%3A%20auto%3B%0A%7D%0Acode%2C%0Akbd%2C%0Apre%20%7B%0Afont%2Dfamily%3A%20monospace%2C%20monospace%3B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ainput%20%7B%0Acolor%3A%20inherit%3B%0Afont%3A%20inherit%3B%0Amargin%3A%200%3B%0A%7D%0Ahtml%20input%5Bdisabled%5D%20%7B%0Acursor%3A%20default%3B%0A%7D%0Ainput%20%7B%0Aline%2Dheight%3A%20normal%3B%0A%7D%0Ainput%5Btype%3D%22checkbox%22%5D%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0Apadding%3A%200%3B%0A%7D%0Atable%20%7B%0Aborder%2Dcollapse%3A%20collapse%3B%0Aborder%2Dspacing%3A%200%3B%0A%7D%0Atd%2C%0Ath%20%7B%0Apadding%3A%200%3B%0A%7D%0A%2A%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0A%7D%0Ainput%20%7B%0Afont%3A%2013px%20%2F%201%2E4%20Helvetica%2C%20arial%2C%20nimbussansl%2C%20liberationsans%2C%20freesans%2C%20clean%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%234078c0%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%2C%0Aa%3Aactive%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200%3B%0Amargin%3A%2015px%200%3B%0Aoverflow%3A%20hidden%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23ddd%3B%0A%7D%0Ahr%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ahr%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%2015px%3B%0Amargin%2Dbottom%3A%2015px%3B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%2030px%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%2021px%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%2016px%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%2011px%3B%0A%7D%0Ablockquote%20%7B%0Amargin%3A%200%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Aol%20ol%2C%0Aul%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Droman%3B%0A%7D%0Aul%20ul%20ol%2C%0Aul%20ol%20ol%2C%0Aol%20ul%20ol%2C%0Aol%20ol%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Dalpha%3B%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0Afont%3A%2012px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0A%7D%0A%2Eselect%3A%3A%2Dms%2Dexpand%20%7B%0Aopacity%3A%200%3B%0A%7D%0A%2Eocticon%20%7B%0Afont%3A%20normal%20normal%20normal%2016px%2F1%20octicons%2Dlink%3B%0Adisplay%3A%20inline%2Dblock%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Drendering%3A%20auto%3B%0A%2Dwebkit%2Dfont%2Dsmoothing%3A%20antialiased%3B%0A%2Dmoz%2Dosx%2Dfont%2Dsmoothing%3A%20grayscale%3B%0A%2Dwebkit%2Duser%2Dselect%3A%20none%3B%0A%2Dmoz%2Duser%2Dselect%3A%20none%3B%0A%2Dms%2Duser%2Dselect%3A%20none%3B%0Auser%2Dselect%3A%20none%3B%0A%7D%0A%2Eocticon%2Dlink%3Abefore%20%7B%0Acontent%3A%20%27%5Cf05c%27%3B%0A%7D%0A%2Emarkdown%2Dbody%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%20%21important%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%20%21important%3B%0A%7D%0Aa%3Anot%28%5Bhref%5D%29%20%7B%0Acolor%3A%20inherit%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0A%2Eanchor%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%2Dright%3A%202px%3B%0Amargin%2Dleft%3A%20%2D18px%3B%0A%7D%0A%2Eanchor%3Afocus%20%7B%0Aoutline%3A%20none%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%201em%3B%0Amargin%2Dbottom%3A%2016px%3B%0Afont%2Dweight%3A%20bold%3B%0Aline%2Dheight%3A%201%2E4%3B%0A%7D%0Ah1%20%2Eocticon%2Dlink%2C%0Ah2%20%2Eocticon%2Dlink%2C%0Ah3%20%2Eocticon%2Dlink%2C%0Ah4%20%2Eocticon%2Dlink%2C%0Ah5%20%2Eocticon%2Dlink%2C%0Ah6%20%2Eocticon%2Dlink%20%7B%0Acolor%3A%20%23000%3B%0Avertical%2Dalign%3A%20middle%3B%0Avisibility%3A%20hidden%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%2C%0Ah2%3Ahover%20%2Eanchor%2C%0Ah3%3Ahover%20%2Eanchor%2C%0Ah4%3Ahover%20%2Eanchor%2C%0Ah5%3Ahover%20%2Eanchor%2C%0Ah6%3Ahover%20%2Eanchor%20%7B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah2%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah3%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah4%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah5%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah6%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%20%7B%0Avisibility%3A%20visible%3B%0A%7D%0Ah1%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%202%2E25em%3B%0Aline%2Dheight%3A%201%2E2%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah1%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah2%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%201%2E75em%3B%0Aline%2Dheight%3A%201%2E225%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah2%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%201%2E5em%3B%0Aline%2Dheight%3A%201%2E43%3B%0A%7D%0Ah3%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%201%2E25em%3B%0A%7D%0Ah4%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ah5%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%201em%3B%0Acolor%3A%20%23777%3B%0A%7D%0Ah6%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ap%2C%0Ablockquote%2C%0Aul%2C%0Aol%2C%0Adl%2C%0Atable%2C%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%204px%3B%0Apadding%3A%200%3B%0Amargin%3A%2016px%200%3B%0Abackground%2Dcolor%3A%20%23e7e7e7%3B%0Aborder%3A%200%20none%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%2Dleft%3A%202em%3B%0A%7D%0Aul%20ul%2C%0Aul%20ol%2C%0Aol%20ol%2C%0Aol%20ul%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Ali%3Ep%20%7B%0Amargin%2Dtop%3A%2016px%3B%0A%7D%0Adl%20%7B%0Apadding%3A%200%3B%0A%7D%0Adl%20dt%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%2016px%3B%0Afont%2Dsize%3A%201em%3B%0Afont%2Dstyle%3A%20italic%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Adl%20dd%20%7B%0Apadding%3A%200%2016px%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ablockquote%20%7B%0Apadding%3A%200%2015px%3B%0Acolor%3A%20%23777%3B%0Aborder%2Dleft%3A%204px%20solid%20%23ddd%3B%0A%7D%0Ablockquote%3E%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Ablockquote%3E%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Atable%20%7B%0Adisplay%3A%20block%3B%0Awidth%3A%20100%25%3B%0Aoverflow%3A%20auto%3B%0Aword%2Dbreak%3A%20normal%3B%0Aword%2Dbreak%3A%20keep%2Dall%3B%0A%7D%0Atable%20th%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Atable%20th%2C%0Atable%20td%20%7B%0Apadding%3A%206px%2013px%3B%0Aborder%3A%201px%20solid%20%23ddd%3B%0A%7D%0Atable%20tr%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Aborder%2Dtop%3A%201px%20solid%20%23ccc%3B%0A%7D%0Atable%20tr%3Anth%2Dchild%282n%29%20%7B%0Abackground%2Dcolor%3A%20%23f8f8f8%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Abackground%2Dcolor%3A%20%23fff%3B%0A%7D%0Acode%20%7B%0Apadding%3A%200%3B%0Apadding%2Dtop%3A%200%2E2em%3B%0Apadding%2Dbottom%3A%200%2E2em%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%2085%25%3B%0Abackground%2Dcolor%3A%20rgba%280%2C0%2C0%2C0%2E04%29%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0Acode%3Abefore%2C%0Acode%3Aafter%20%7B%0Aletter%2Dspacing%3A%20%2D0%2E2em%3B%0Acontent%3A%20%22%5C00a0%22%3B%0A%7D%0Apre%3Ecode%20%7B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%20100%25%3B%0Aword%2Dbreak%3A%20normal%3B%0Awhite%2Dspace%3A%20pre%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0A%2Ehighlight%20%7B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0A%2Ehighlight%20pre%2C%0Apre%20%7B%0Apadding%3A%2016px%3B%0Aoverflow%3A%20auto%3B%0Afont%2Dsize%3A%2085%25%3B%0Aline%2Dheight%3A%201%2E45%3B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0A%2Ehighlight%20pre%20%7B%0Amargin%2Dbottom%3A%200%3B%0Aword%2Dbreak%3A%20normal%3B%0A%7D%0Apre%20%7B%0Aword%2Dwrap%3A%20normal%3B%0A%7D%0Apre%20code%20%7B%0Adisplay%3A%20inline%3B%0Amax%2Dwidth%3A%20initial%3B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Aoverflow%3A%20initial%3B%0Aline%2Dheight%3A%20inherit%3B%0Aword%2Dwrap%3A%20normal%3B%0Abackground%2Dcolor%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0Apre%20code%3Abefore%2C%0Apre%20code%3Aafter%20%7B%0Acontent%3A%20normal%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%2Dsize%3A%2011px%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Epl%2Dc%20%7B%0Acolor%3A%20%23969896%3B%0A%7D%0A%2Epl%2Dc1%2C%0A%2Epl%2Ds%20%2Epl%2Dv%20%7B%0Acolor%3A%20%230086b3%3B%0A%7D%0A%2Epl%2De%2C%0A%2Epl%2Den%20%7B%0Acolor%3A%20%23795da3%3B%0A%7D%0A%2Epl%2Ds%20%2Epl%2Ds1%2C%0A%2Epl%2Dsmi%20%7B%0Acolor%3A%20%23333%3B%0A%7D%0A%2Epl%2Dent%20%7B%0Acolor%3A%20%2363a35c%3B%0A%7D%0A%2Epl%2Dk%20%7B%0Acolor%3A%20%23a71d5d%3B%0A%7D%0A%2Epl%2Dpds%2C%0A%2Epl%2Ds%2C%0A%2Epl%2Ds%20%2Epl%2Dpse%20%2Epl%2Ds1%2C%0A%2Epl%2Dsr%2C%0A%2Epl%2Dsr%20%2Epl%2Dcce%2C%0A%2Epl%2Dsr%20%2Epl%2Dsra%2C%0A%2Epl%2Dsr%20%2Epl%2Dsre%20%7B%0Acolor%3A%20%23183691%3B%0A%7D%0A%2Epl%2Dv%20%7B%0Acolor%3A%20%23ed6a43%3B%0A%7D%0A%2Epl%2Did%20%7B%0Acolor%3A%20%23b52a1d%3B%0A%7D%0A%2Epl%2Dii%20%7B%0Abackground%2Dcolor%3A%20%23b52a1d%3B%0Acolor%3A%20%23f8f8f8%3B%0A%7D%0A%2Epl%2Dsr%20%2Epl%2Dcce%20%7B%0Acolor%3A%20%2363a35c%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dml%20%7B%0Acolor%3A%20%23693a17%3B%0A%7D%0A%2Epl%2Dmh%2C%0A%2Epl%2Dmh%20%2Epl%2Den%2C%0A%2Epl%2Dms%20%7B%0Acolor%3A%20%231d3e81%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmq%20%7B%0Acolor%3A%20%23008080%3B%0A%7D%0A%2Epl%2Dmi%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dstyle%3A%20italic%3B%0A%7D%0A%2Epl%2Dmb%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmd%20%7B%0Abackground%2Dcolor%3A%20%23ffecec%3B%0Acolor%3A%20%23bd2c00%3B%0A%7D%0A%2Epl%2Dmi1%20%7B%0Abackground%2Dcolor%3A%20%23eaffea%3B%0Acolor%3A%20%2355a532%3B%0A%7D%0A%2Epl%2Dmdr%20%7B%0Acolor%3A%20%23795da3%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmo%20%7B%0Acolor%3A%20%231d3e81%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%3A%2011px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20%7B%0Alist%2Dstyle%2Dtype%3A%20none%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%2B%2Etask%2Dlist%2Ditem%20%7B%0Amargin%2Dtop%3A%203px%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20input%20%7B%0Amargin%3A%200%200%2E35em%200%2E25em%20%2D1%2E6em%3B%0Avertical%2Dalign%3A%20middle%3B%0A%7D%0A%3Achecked%2B%2Eradio%2Dlabel%20%7B%0Az%2Dindex%3A%201%3B%0Aposition%3A%20relative%3B%0Aborder%2Dcolor%3A%20%234078c0%3B%0A%7D%0A%2EsourceLine%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0A%7D%0Acode%20%2Ekw%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Edt%20%7B%20color%3A%20%23ed6a43%3B%20%7D%0Acode%20%2Edv%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Ebn%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Efl%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Ech%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%2Est%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Eco%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%2Eot%20%7B%20color%3A%20%230086b3%3B%20%7D%0Acode%20%2Eal%20%7B%20color%3A%20%23a61717%3B%20%7D%0Acode%20%2Efu%20%7B%20color%3A%20%2363a35c%3B%20%7D%0Acode%20%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%0Acode%20%2Ewa%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Ecn%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Esc%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Evs%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Ess%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%2Eim%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eva%20%7Bcolor%3A%20%23008080%3B%20%7D%0Acode%20%2Ecf%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eop%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Ebu%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Eex%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%2Epp%20%7B%20color%3A%20%23999999%3B%20%7D%0Acode%20%2Eat%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Edo%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%2Ean%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Ecv%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%2Ein%20%7B%20color%3A%20%23008080%3B%20%7D%0A">
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="conditional-and-joint-analysis">Conditional and Joint Analysis</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span></code></pre></div>
<h1 id="methods">Methods</h1>
<p>We ran a <a href="https://www.nature.com/articles/ng.2213">conditional and joint analysis</a> using <a href="https://cnsgenomics.com/software/gcta/#COJO">GCTA</a> to refine the list of independent loci.</p>
<ul>
<li>Final meta-analysed SNPs from the Ricopili pipeline were used.</li>
<li>Ricopili was used for initial clumping with index SNPs identified with <em>p</em>&lt; 10^{-4} and <em>r</em><sup>2</sup>&lt; 0.1 within 3000kb windows. The extended MHC region was clumped as a single region.</li>
<li>Sumstats were filtered for MAF &gt;= 0.01 and INFO &gt; 0.6</li>
<li>Regions with a genome-wide significant SNP (p &lt; 5-e8) were identified from the clumped results. Regions within 50kb of each other were merged.</li>
<li>SNPs from these regions were extracted filtered to unrelated of self- and genotype-identified European ancestry participants from UK Biobank.</li>
<li>A conditional analysis was performed on each region using the filtered sumstats superimposed on the UK Biobank LD structure.</li>
</ul>
<h2 id="previous-sumstats">Previous sumstats</h2>
<p>Variants from <a href="https://www.nature.com/articles/s41593-018-%200326-7">Howard et al 2019</a> (remove first and last rows with table captions) and <a href="https://doi.org/10.1038/s41593-021-00860-2">Levey et al 2021</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>howard <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>howard, <span class="at">skip=</span><span class="dv">2</span>, <span class="at">n_max=</span><span class="dv">102</span>)</span></code></pre></div>
<pre><code>## New names:
## * `Odds Ratio` -&gt; `Odds Ratio...8`
## * `Lower 95% Confidence Interval` -&gt; `Lower 95% Confidence Interval...9`
## * `Upper 95% Confidence Interval` -&gt; `Upper 95% Confidence Interval...10`
## * `Log(Odds Ratio)` -&gt; `Log(Odds Ratio)...11`
## * `Standard error of the Log(Odds Ratio)` -&gt; `Standard error of the Log(Odds Ratio)...12`
## * ...</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>levey <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>levey, <span class="at">col_types=</span><span class="fu">cols</span>(<span class="at">CHR.BP=</span><span class="fu">col_character</span>()))</span></code></pre></div>
<h1 id="results">Results</h1>
<h2 id="cojo-snp-and-region-counts">COJO SNP and region counts</h2>
<p>List of clumped and COJO SNPs and regions</p>
<ul>
<li>Sumstats: results/cojo/daner_pgc_mdd_full_eur_hg19_v3.49.24.05.rp.qc.gz</li>
<li>Clump file: results/distribution/daner_pgc_mdd_full_eur_hg19_v3.49.24.05.rp.gz.p4.clump.areator.sorted.1mhc</li>
<li>COJO regions: 549</li>
<li>Clumped SNPs: 738</li>
<li>COJO Selected SNPs: 594</li>
<li>Singleton regions: 63</li>
<li>COJO+Clump SNPs: 606</li>
<li>COJO Final SNPs: 543</li>
<li>COJO Final SNPs p &lt;= 5e-8, pJ &lt;= 5e-8: 512</li>
<li>COJO Final SNPs p &gt; 5e-8, pJ &lt;= 5e-8: 29</li>
</ul>
<p>Load list of COJO SNPs</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>cojo <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>cojo)</span></code></pre></div>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   SNP = col_character(),
##   A1 = col_character(),
##   A2 = col_character(),
##   Direction = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<h2 id="clumped-results">Clumped results</h2>
<p>Clumped results from Ricopili and METACARPA for comparison</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rp <span class="ot">&lt;-</span> <span class="fu">read_table2</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>rp_clump) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(P <span class="sc">&lt;=</span> <span class="fl">5e-8</span>)</span></code></pre></div>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   SNP = col_character(),
##   A1A2 = col_character(),
##   `(Nca,Nco,Neff)Dir` = col_character(),
##   ngt = col_character(),
##   `LD-friends(0.1).p0.001` = col_character(),
##   `LD-friends(0.6).p0.001` = col_character(),
##   gwas_catalog_span.6 = col_character(),
##   `genes.6.50kb(dist2index)` = col_character(),
##   N.genes.6.50kb = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>mc <span class="ot">&lt;-</span> <span class="fu">read_table2</span>(snakemake<span class="sc">@</span>input<span class="sc">$</span>mc_clump) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(P <span class="sc">&lt;=</span> <span class="fl">5e-8</span>)</span></code></pre></div>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   SNP = col_character(),
##   A1A2 = col_character(),
##   ngt = col_character(),
##   `LD-friends(0.1).p0.001` = col_character(),
##   `LD-friends(0.6).p0.001` = col_character(),
##   gwas_catalog_span.6 = col_character(),
##   `genes.6.50kb(dist2index)` = col_character(),
##   N.genes.6.50kb = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<h2 id="genomic-ranges">Genomic ranges</h2>
<p>Construct genomic range objects so that SNPs and regions can be intersected and compared.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>cojo_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(cojo, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>range.left, <span class="at">end=</span>range.right), <span class="at">SNP=</span>SNP))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>rp_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(rp, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>range.left, <span class="at">end=</span>range.right), <span class="at">SNP=</span>SNP))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>mc_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(mc, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>range.left, <span class="at">end=</span>range.right), <span class="at">SNP=</span>SNP))</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>howard_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(howard, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>Chromosome, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span><span class="st">`</span><span class="at">Postion (bp)</span><span class="st">`</span>, <span class="at">width=</span><span class="dv">1</span>)))</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>levey_gr <span class="ot">&lt;-</span> <span class="fu">with</span>(levey, <span class="fu">GRanges</span>(<span class="at">seqnames=</span>CHR, <span class="at">ranges=</span><span class="fu">IRanges</span>(<span class="at">start=</span>BP, <span class="at">width=</span><span class="dv">1</span>)))</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>rp_mc_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(rp_gr, mc_gr)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>rp_and_mc <span class="ot">&lt;-</span> rp <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(rp_mc_overlaps<span class="sc">@</span>from))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>rp_not_mc <span class="ot">&lt;-</span> rp <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="sc">-</span><span class="fu">unique</span>(rp_mc_overlaps<span class="sc">@</span>from))</span></code></pre></div>
<h2 id="gwas-catalog">GWAS catalog</h2>
<p>Tally entries in the GWAS catalog for each region</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>parse_catalog_entry <span class="ot">&lt;-</span> <span class="cf">function</span>(catalog_entry) {</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># extract LD and RSID from first element</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    ld_snp_match <span class="ot">&lt;-</span> <span class="fu">str_match</span>(catalog_entry[<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)(rs[[:digit:]]+)&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    ld <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(ld_snp_match[,<span class="dv">2</span>])</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    snp <span class="ot">&lt;-</span> ld_snp_match[,<span class="dv">3</span>]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># parse rest of elements into phenotype(PubMed ID)(P-value)</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    phenotype_matches <span class="ot">&lt;-</span> <span class="fu">str_match</span>(catalog_entry[<span class="sc">-</span><span class="dv">1</span>], <span class="st">&quot;(.+)</span><span class="sc">\\</span><span class="st">(([[:digit:]]+)</span><span class="sc">\\</span><span class="st">)</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)&quot;</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    phenotypes <span class="ot">&lt;-</span> phenotype_matches[,<span class="dv">2</span>]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    pubmeds <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenotype_matches[,<span class="dv">3</span>])</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    p_values <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenotype_matches[,<span class="dv">4</span>])</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">data.frame</span>(ld, <span class="at">catalogSNP=</span>snp, <span class="at">phenotype=</span>phenotypes, <span class="at">pubmed_id=</span>pubmeds, <span class="at">P=</span>p_values))</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>rp_gwas_catalog_entries <span class="ot">&lt;-</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>plyr<span class="sc">::</span><span class="fu">ddply</span>(<span class="fu">filter</span>(rp, gwas_catalog_span<span class="fl">.6</span> <span class="sc">!=</span> <span class="st">&#39;-&#39;</span>), <span class="sc">~</span>SNP, <span class="cf">function</span>(rp_entry) {</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get GWAS catalog cell</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    gwas_catalog <span class="ot">&lt;-</span> rp_entry<span class="sc">$</span>gwas_catalog_span<span class="fl">.6</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># split out into catalog entries (separated by /, removing first empty element)</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    catalog_entries <span class="ot">&lt;-</span> <span class="fu">str_split</span>(gwas_catalog, <span class="st">&quot;/&quot;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>    catalog_entries_complete <span class="ot">&lt;-</span> catalog_entries[<span class="fu">which</span>(catalog_entries <span class="sc">!=</span> <span class="st">&quot;&quot;</span>)]</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># split SNP entries by &quot;;&quot;</span></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>    catalog_entries_list <span class="ot">&lt;-</span> <span class="fu">str_split</span>(catalog_entries_complete, <span class="st">&quot;;&quot;</span>)</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># parse each entry</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>    catalog_entries_df <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">ldply</span>(catalog_entries_list, parse_catalog_entry)</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(catalog_entries_df)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<h1 id="parse-gene-list">Parse gene list</h1>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="ot">&lt;-</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>plyr<span class="sc">::</span><span class="fu">ddply</span>(<span class="fu">filter</span>(rp, <span class="st">`</span><span class="at">genes.6.50kb(dist2index)</span><span class="st">`</span> <span class="sc">!=</span> <span class="st">&#39;-&#39;</span>), <span class="sc">~</span>SNP, <span class="cf">function</span>(rp_entry) {</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    genes_dist <span class="ot">&lt;-</span> rp_entry<span class="sc">$</span><span class="st">`</span><span class="at">genes.6.50kb(dist2index)</span><span class="st">`</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    genes_dist_list <span class="ot">&lt;-</span> <span class="fu">str_split</span>(genes_dist, <span class="st">&#39;,&#39;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    genes_dist_match <span class="ot">&lt;-</span> <span class="fu">str_match</span>(genes_dist_list, <span class="st">&quot;(.+)</span><span class="sc">\\</span><span class="st">((.+)</span><span class="sc">\\</span><span class="st">)&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="at">gene=</span>genes_dist_match[,<span class="dv">2</span>], <span class="at">dist2index=</span><span class="fu">as.numeric</span>(genes_dist_match[,<span class="dv">3</span>])))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<h2 id="comparison-to-previous-findings">Comparison to previous findings</h2>
<p>Find which COJO regions overlap with Howard</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>cojo_howard_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(cojo_gr, howard_gr)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>cojo_howard_overlaps</span></code></pre></div>
<pre><code>## Hits object with 128 hits and 0 metadata columns:
##         queryHits subjectHits
##         &lt;integer&gt;   &lt;integer&gt;
##     [1]         1           1
##     [2]         5           2
##     [3]         6           3
##     [4]         9           4
##     [5]        10           4
##     ...       ...         ...
##   [124]       519          97
##   [125]       522          98
##   [126]       527         100
##   [127]       528         101
##   [128]       541         102
##   -------
##   queryLength: 543 / subjectLength: 102</code></pre>
<p>Count number of regions in Howard that overlap:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>howard <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_howard_overlaps<span class="sc">@</span>to)) <span class="sc">%&gt;%</span> <span class="fu">count</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 × 1
##       n
##   &lt;int&gt;
## 1    85</code></pre>
<p>Find which COJO regions overlap with Levey</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>cojo_levey_overlaps <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(cojo_gr, levey_gr)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>cojo_levey_overlaps</span></code></pre></div>
<pre><code>## Hits object with 270 hits and 0 metadata columns:
##         queryHits subjectHits
##         &lt;integer&gt;   &lt;integer&gt;
##     [1]         1         110
##     [2]         5           2
##     [3]         6          65
##     [4]         9          77
##     [5]         9          26
##     ...       ...         ...
##   [266]       531         154
##   [267]       532         154
##   [268]       538         120
##   [269]       541         153
##   [270]       542         211
##   -------
##   queryLength: 543 / subjectLength: 223</code></pre>
<p>Count number of regions in Levey that overlap:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>levey <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_levey_overlaps<span class="sc">@</span>to)) <span class="sc">%&gt;%</span> <span class="fu">count</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 × 1
##       n
##   &lt;int&gt;
## 1   186</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>cojo_known <span class="ot">&lt;-</span> cojo <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(cojo_levey_overlaps<span class="sc">@</span>from))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>catalog_known <span class="ot">&lt;-</span> rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_known<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>catalog_known</span></code></pre></div>
<pre><code>## # A tibble: 76 × 2
##    phenotype                     n
##    &lt;chr&gt;                     &lt;int&gt;
##  1 Schizophrenia                26
##  2 Neuroticism                  18
##  3 Depression                   17
##  4 Intelligence_(MTAG)          16
##  5 Depressive_symp...           15
##  6 Body_mass_index              12
##  7 Autism_spectrum...           11
##  8 Depressive_symptoms          10
##  9 Neuroticism_(MTAG)           10
## 10 Major_depressive_disorder     9
## # … with 66 more rows</code></pre>
<p>Newly discovered regions</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>cojo_new <span class="ot">&lt;-</span> cojo <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="fu">unique</span>(<span class="sc">-</span>cojo_levey_overlaps<span class="sc">@</span>from)) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(P)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>cojo_new</span></code></pre></div>
<pre><code>## # A tibble: 351 × 28
##    region snp_idx   CHR SNP               BP A1    A2    FRQ_A_511755 FRQ_U_3072803
##     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;        &lt;dbl&gt;         &lt;dbl&gt;
##  1    476       1    20 rs34966255  51193862 C     T            0.18          0.183
##  2     50       2     2 rs951807    60513389 T     C            0.404         0.405
##  3    182       1     5 rs1993739  153215007 T     C            0.166         0.171
##  4     50       1     2 rs10490064  60163127 A     G            0.367         0.38 
##  5    215       1     7 rs1114780    3551788 C     T            0.237         0.242
##  6    243       1     7 rs7797141  115052164 A     G            0.274         0.266
##  7    174       1     5 rs7726836  120065919 T     A            0.294         0.285
##  8    354       1    12 rs2363585   60791165 A     T            0.375         0.383
##  9    368       1    12 rs7962128  121907336 A     G            0.455         0.46 
## 10     80       1     2 rs13418032 198413692 G     C            0.334         0.323
## # … with 341 more rows, and 19 more variables: INFO &lt;dbl&gt;, OR &lt;dbl&gt;, SE &lt;dbl&gt;,
## #   P &lt;dbl&gt;, ngt &lt;dbl&gt;, Direction &lt;chr&gt;, HetISqt &lt;dbl&gt;, HetDf &lt;dbl&gt;,
## #   HetPVa &lt;dbl&gt;, Nca &lt;dbl&gt;, Nco &lt;dbl&gt;, Neff_half &lt;dbl&gt;, bJ &lt;dbl&gt;, bJ_se &lt;dbl&gt;,
## #   pJ &lt;dbl&gt;, LD_r &lt;dbl&gt;, range.left &lt;dbl&gt;, range.right &lt;dbl&gt;, n_sig_snps &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_new<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n)) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span>phenotype <span class="sc">%in%</span> catalog_known<span class="sc">$</span>phenotype)</span></code></pre></div>
<pre><code>## # A tibble: 52 × 2
##    phenotype                    n
##    &lt;chr&gt;                    &lt;int&gt;
##  1 Height                       4
##  2 Diastolic_blood_pressure     3
##  3 Morning_vs._eve...           3
##  4 Adiponectin_levels           2
##  5 Blood_osmolalit...           2
##  6 Hip_circumferen...           2
##  7 LDL_cholesterol_levels       2
##  8 Lobe_attachment...           2
##  9 Lymphocyte_perc...           2
## 10 Mean_corpuscular_volume      2
## # … with 42 more rows</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> cojo_new<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(SNP) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(dist2index <span class="sc">==</span> <span class="fu">min</span>(dist2index)) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="fu">select</span>(gene) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>()</span></code></pre></div>
<pre><code>## # A tibble: 261 × 1
##    gene    
##    &lt;chr&gt;   
##  1 PPP3CA  
##  2 MIR1255A
##  3 FLJ20021
##  4 NISCH   
##  5 STAB1   
##  6 NT5DC2  
##  7 SMIM4   
##  8 PBRM1   
##  9 NTRK3   
## 10 GTF2IRD1
## # … with 251 more rows</code></pre>
<h2 id="comparison-between-pre-cojo-and-post-cojo">Comparison between pre-COJO and post-COJO</h2>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span>SNP <span class="sc">%in%</span> cojo<span class="sc">$</span>SNP)</span></code></pre></div>
<pre><code>## # A tibble: 1,108 × 3
##    SNP        gene       dist2index
##    &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
##  1 rs10064584 LINC02240           0
##  2 rs10089596 SGCZ                0
##  3 rs10165927 PLA2R1              0
##  4 rs10165927 ITGB6               0
##  5 rs10170048 UBE2F-SCLY          0
##  6 rs10170048 SCLY                0
##  7 rs10170048 ESPNL               0
##  8 rs10170048 KLHL30              0
##  9 rs10170048 ERFE                0
## 10 rs10170048 ILKAP               0
## # … with 1,098 more rows</code></pre>
<h2 id="comparison-between-ricopili-and-metacarpa-clumped-results">Comparison between Ricopili and METACARPA clumped results</h2>
<p>Shared regions and regions not significant in the METACARPA analysis.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>rp_and_mc_phenotypes <span class="ot">&lt;-</span> rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> rp_and_mc<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>rp_not_mc_phenotypes <span class="ot">&lt;-</span> rp_gwas_catalog_entries <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> rp_not_mc<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">count</span>(phenotype) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>rp_not_mc_phenotypes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span>phenotype <span class="sc">%in%</span> rp_and_mc_phenotypes<span class="sc">$</span>phenotype)</span></code></pre></div>
<pre><code>## # A tibble: 10 × 2
##    phenotype                    n
##    &lt;chr&gt;                    &lt;int&gt;
##  1 Menopause_(age_at_onset)     3
##  2 Mean_corpuscular_volume      2
##  3 Visceral_adipos...           2
##  4 Adenocarcinoma               1
##  5 Hemoglobin                   1
##  6 Lipoprotein-ass...           1
##  7 Mean_corpuscula...           1
##  8 Monobrow                     1
##  9 Prostate_cancer              1
## 10 White_blood_cel...           1</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(SNP) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(dist2index <span class="sc">==</span> <span class="fu">min</span>(dist2index)) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="fu">select</span>(gene) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>()</span></code></pre></div>
<pre><code>## # A tibble: 790 × 1
##    gene     
##    &lt;chr&gt;    
##  1 PPP3CA   
##  2 MIR1255A 
##  3 FLJ20021 
##  4 CD14     
##  5 LINC02240
##  6 SGCZ     
##  7 NISCH    
##  8 STAB1    
##  9 NT5DC2   
## 10 SMIM4    
## # … with 780 more rows</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> rp_and_mc<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(SNP) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(dist2index <span class="sc">==</span> <span class="fu">min</span>(dist2index)) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="fu">select</span>(gene) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>()</span></code></pre></div>
<pre><code>## # A tibble: 679 × 1
##    gene     
##    &lt;chr&gt;    
##  1 PPP3CA   
##  2 MIR1255A 
##  3 FLJ20021 
##  4 CD14     
##  5 LINC02240
##  6 SGCZ     
##  7 NISCH    
##  8 STAB1    
##  9 NT5DC2   
## 10 SMIM4    
## # … with 669 more rows</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>rp_genes_dist <span class="sc">%&gt;%</span> <span class="fu">filter</span>(SNP <span class="sc">%in%</span> rp_not_mc<span class="sc">$</span>SNP) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(SNP) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(dist2index <span class="sc">==</span> <span class="fu">min</span>(dist2index)) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="fu">select</span>(gene) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>()</span></code></pre></div>
<pre><code>## # A tibble: 118 × 1
##    gene      
##    &lt;chr&gt;     
##  1 UBE2F-SCLY
##  2 SCLY      
##  3 ESPNL     
##  4 KLHL30    
##  5 ERFE      
##  6 ILKAP     
##  7 SHTN1     
##  8 VAX1      
##  9 MIR3663HG 
## 10 MIR3663   
## # … with 108 more rows</code></pre>

</body>
</html>
