---
title: Structure Meta-Analysis with GenomicSEM
output: github_document
---

Summary statistics were grouped based on phenotype, then meta-analysed:

- `Clin`: Clinical assessment
- `EHR`: Electronic health records
- `Quest`: Questionnaire
- `SelfRep`: Single-item self report

The LDSC covariance structure of the four MDD phenotype was calcuted using [GenomicSEM](https://github.com/GenomicSEM/GenomicSEM).

```{r libraries, results='hide', message=FALSE, warning=FALSE}

library(GenomicSEM)
library(readr)
library(corrplot)
library(dplyr)

save.image()

```

Read in the covariance structure and the LDSC 

```{r covstruct}

mdd_covstruct <- dget(snakemake@input$covstruct)

```

LDSC statistics

```{r ldsc}
ldsc_table <- read_tsv(snakemake@input$ldsc_table)
knitr::kable(ldsc_table)
```

Genetic correlations

```{r rg, results='hide'}

corrplot(cov2cor(mdd_covstruct$S), method='number')

```

## Common factor model

```{r model}

common.model <- "A =~ NA*Clin + EHR + Quest + SelfRep
A ~~ 1*A"

common.fit <- usermodel(covstruc=mdd_covstruct, estimation='DWLS', model=common.model)

```

```{r fit}

knitr::kable(common.fit$modelfit)

```

```{r results}

common.fit.results <- common.fit$results %>%
    mutate(Unstand_SE=as.numeric(Unstand_SE),
           STD_Genotype_SE=as.numeric(STD_Genotype_SE),
           p_value=if_else(p_value == '< 5e-300', true=0, false=as.numeric(p_value)))

knitr::kable(common.fit.results, digits=4)

```

## GWAS-by-subtraction

Fit a [GWAS-by-subtraction](https://rpubs.com/MichelNivard/565885) to decompose variance into that of minimally-phenotyped depression shared with maximal phenotypes versus specific to maximal phenotypes.

```{r gbs_model}

md.model <- "MIN =~ NA*SelfRep + Clin + EHR + Quest
MAX =~ NA*Clin + EHR + Quest

MIN ~~ 1*MIN
MAX ~~ 1*MAX
MIN ~~ 0*MAX

SelfRep ~~ 0*Clin + 0*EHR + 0*Quest
SelfRep ~~ 0*SelfRep"

```

```{r gbs_fit}
md.fit <- usermodel(covstruc=mdd_covstruct, estimation='DWLS', model=md.model)
```

```{r gbs_results}

md.fit.results <- md.fit$results %>%
mutate(Unstand_SE=as.numeric(Unstand_SE),
       STD_Genotype_SE=as.numeric(STD_Genotype_SE),
       p_value=if_else(p_value == '< 5e-300', true=0, false=as.numeric(p_value)))

knitr::kable(md.fit.results, digits=4)

```