---
title: Structure Meta-Analysis with GenomicSEM
output: github_document
---

Summary statistics were grouped based on phenotype, then meta-analysed:

- `Clin`: Clinical assessment
- `EHR`: Electronic health records
- `Quest`: Questionnaire
- `SelfRep`: Single-item self report

The LDSC covariance structure of the four MDD phenotype was calcuted using [GenomicSEM](https://github.com/GenomicSEM/GenomicSEM).

```{r libraries, results='hide', message=FALSE, warning=FALSE}
library(GenomicSEM)
library(readr)
library(corrplot)
library(miamiplot)
library(dplyr)
```

Read in the covariance structure and the LDSC 

```{r covstruct}

mdd_covstruct <- dget(snakemake@input$covstruct)

```

LDSC statistics

```{r ldsc}
ldsc_table <- read_tsv(snakemake@input$ldsc_table)
knitr::kable(ldsc_table)
```

Genetic correlations

```{r rg, results='hide'}

corrplot(cov2cor(mdd_covstruct$S), method = "number")

```

## Common factor model

```{r model}

common.model <- "
MD =~ NA*Clin + EHR + Quest + SelfRep
MD ~~ 1*MD
"

common.fit <- usermodel(covstruc = mdd_covstruct,
                        estimation = "DWLS",
                        model = common.model,
                        imp_cov=TRUE)

```

```{r fit}

knitr::kable(common.fit$modelfit)

```

```{r results}

common.fit.results <- common.fit$results %>%
    mutate(Unstand_SE = as.numeric(Unstand_SE),
           STD_Genotype_SE = as.numeric(STD_Genotype_SE),
           p_value=if_else(p_value == "< 5e-300",
                           true = 0,
                           false = as.numeric(p_value))) %>%
                           mutate(STD_All_SE=STD_Genotype_SE*STD_All/STD_Genotype) 

knitr::kable(select(common.fit.results,
                    lhs, op, rhs,
                    STD_All, STD_All_SE,
                    p_value),
                    digits = 4)

```

Model-implied correlations

```{r imp_cov}

cov2cor(common.fit$resid_cov$`Model Implied Covariance Matrix`)

```

### Common factor GWAS

Miami plot of P values and Q values

```{r gwas, dpi=300}

common.gwas <- read_tsv(snakemake@input$commonfactor_eur)

p.gwas <- common.gwas %>% transmute(rsid=SNP, chr=CHR, pos=BP, pval=Pval_Estimate, P="P")
q.gwas <- common.gwas %>% transmute(rsid=SNP, chr=CHR, pos=BP, pval=Q_pval, P="Q")

ggmiami(data=bind_rows(p.gwas, q.gwas) %>% filter(pval <= 0.001) %>% as.data.frame(),
        split_by="P", split_at="P", p="pval",
        upper_ylab="Common Factor GWAS",
        lower_ylab="Heterogeneity Q")

```

## Anchor model

Fit a one-factor model where the factor is anchored to the clinical phenotype and explains all of its variance.

```{r mdd_model}

mdd.model <-  "
MDD =~ 1*Clin + EHR + Quest + SelfRep
MDD ~~ MDD
Clin ~~ 0*Clin
"

```

```{r mdd_fit}
mdd.fit <- usermodel(covstruc=mdd_covstruct, estimation='DWLS', model=mdd.model, imp_cov=TRUE)
```


```{r mdd_imp_cov}

cov2cor(mdd.fit$resid_cov$`Model Implied Covariance Matrix`)

```

```{r mdd_results}

mdd.fit.results <- mdd.fit$results %>%
mutate(Unstand_SE=as.numeric(Unstand_SE),
       STD_Genotype_SE=as.numeric(STD_Genotype_SE),
       p_value=if_else(p_value == '< 5e-300', true=0, false=as.numeric(p_value))) %>%
mutate(STD_All_SE=STD_Genotype_SE*STD_All/STD_Genotype) 

knitr::kable(select(mdd.fit.results,
    lhs, op, rhs,
    STD_All, STD_All_SE,
    p_value),
    digits = 4)

```