# Analysis Plans

Plans for analyses to include in primary:

## Meta-analysis

- GWAS of genotyped cohorts + meta-analysis of same
- Meta-analysis GWAS of genotyped + summary statistics cohorts
  - Inverse-variance weighted ([Ricopili](https://sites.google.com/a/broadinstitute.org/ricopili/))
  - Structural meta analysis based on phenotyping (eHR – self declared – interview based) ([MTAG](https://github.com/omeed-maghzian/mtag)/[GenomicSEM](https://github.com/MichelNivard/GenomicSEM))
  - Final primary versions:
    - Full meta (PGC only)
    - noUKBB (PGC only)
    - top 10k (Public)
    - no23am (Public)
    - noUKBB / no23am (Public)
- Gene based analysis
 
# Omics
 
- Positional-mapping
- Expression based mapping
- Hi-C (chromosome conformation) ([H-MAGMA](https://github.com/thewonlab/H-MAGMA))
- Other partitioned heritability / gene set enrichment testing
- cell type analysis 
- eQTL
- [S-PrediXcan](https://github.com/hakyimlab/MetaXcan)
- [SMR](https://cnsgenomics.com/software/smr/#Overview) (summary-data-based Mendelian Randomization)
 
## Prediction
 
- SBayesR - estimate joint effect of SNPs / single threshold PRS ([GCTB](https://cnsgenomics.com/software/gctb/#SummaryBayesianAlphabet))
- Replication (or possible meta) in other cohorts
- Leave one out PRS – to be provided to each group
- PheWAS (GenScot and UKB Imaging [or all UKB])
- Predict into other UKB ancestries
