library(dplyr)
library(data.table)
library(TwoSampleMR)
library(optparse)
library(pbapply)
library(readr)

# Load arguments ----------------------------------------------------------
# Environment: R 3.6.1

parse <- OptionParser()

option_list <- list(
   make_option('--SNPlist', type='character', help="SNP-CpG list to obtain", action='store'),
   make_option('--phenolist', type='character', help="SNP-CpG list to obtain", action='store'),
   make_option('--gwas', type='character', help="GWAS sumstats", action='store'),
   make_option('--interexp', type='character', 
               help='Folder for intermediate files generated by processing exposure data', action='store'),
   make_option('--ExposureDat', type='character', default = "",help="Location of exposure data (OutcomeProcess==T)", action='store'),
   make_option('--fNeale', type='character', default = "data/MR/MR_sumstats/Neale/variants.tsv.gz",
               help="Location of info file for Neale lab", action='store'),
   make_option('--fBIG40', type='character', default = "data/MR/MR_sumstats/BIG40/variants.txt.gz",
               help="Location of info file for Neale lab", action='store'),
   make_option('--phenoSummary', type='character', default = "data/MR/pheno_gwas_n.rds",
               help="Location of info file for Neale lab", action='store')
)

args = commandArgs(trailingOnly=TRUE)

opt <- parse_args(OptionParser(option_list=option_list), args=args)
f.snp=opt$SNPlist
f.pheno=opt$phenolist
gwas.f_and_label = opt$gwas
f.exposure = opt$ExposureDat
d.output=opt$interexp


f.Neale = opt$fNeal
f.big40 = opt$fBIG40
f.nsummary = opt$phenoSummary


# f.snp='data/MR/mvmr_ivwT_snp.tsv'
# f.pheno='data/MR/mvmr_ivwT_pheno.tsv'
# gwas.f_and_label= 'data/MR/pheno_gwas_forMR.rds'
# f.exposure='data/MR/MR_InterFiles/exposure_stats/MDD.exposure_dat'
# d.output='data/MR/MR_InterFiles_mvmr'
# f.Neale='data/MR/MR_sumstats/Neale/variants.tsv.gz'
# f.big40='data/MR/MR_sumstats/BIG40/variants.txt.gz'
# f.nsummary='data/MR/pheno_gwas_n.rds'

# Basic settings ----------------------------------------------------------

# SNP and trait lists
ls.snp_include = fread(f.snp,header=F) %>% .$V1
ls.pheno = fread(f.pheno,header=F) %>% .$V1

# inputs
inputs.gwas = readRDS(gwas.f_and_label) %>% 
  .[.$field_tag %in% ls.pheno,]

### Load ref data
# Neale gwas
ref.neale = fread(f.Neale) %>% 
  filter(info>0.8,minor_AF>0.0005) %>% 
  select(variant,A1=ref,A2=alt,rsid,info,freq=AF,chr,pos,n=n_called)
# BIG40 IDP gwas
ref.big40 = fread(f.big40) 
# Phenotype N
ref.n = readRDS(f.nsummary)


# Define functions --------------------------------------------------------

process_gwas <- function(tmp.input,tmp_process.exposure,tmp_process.outcome,ls.snp=NA){
  tmp.fname = tmp.input$file_loc
  target.colname = tmp.input$field_tag
  tmp.type = tmp.input$loading_type
  
  
  target.pheno = target.colname %>% strsplit(.,'.',fixed = T) %>% unlist %>% head(n=2) %>%
    tail(n=1)
  
  tmp.n = ref.n$n[ref.n$phenotype==target.pheno] %>% .[1]
  
  
  ################ Create a list that do not need to process #####################
  # ls.processed = list.files(d.output,pattern = '_dat$',full.names = T,recursive = T) %>% 
  #   basename %>% 
  #   gsub('.exposure_dat','',.,fixed = T) %>% 
  #   gsub('.outcome_dat','',.,fixed = T)
  # 
  # ls.running = read_tsv('data/MR/running_prepare_expo_outc',col_names = F) %>% as.data.frame %>% 
  #   .$X1 %>% as.character %>% as.vector
  # 
  # if(file.exists(paste0(d.output,'/Instrument_info.tsv'))){
  #   ls.no_instrument = read_tsv(paste0(d.output,'/Instrument_info.tsv'),col_names = F) %>% 
  #     as.data.frame %>%  
  #     .$X1 
  # }else{ls.no_instrument=NA}
  # 
  # ls.avoid = c(ls.processed,ls.running,ls.no_instrument)
  # 
  # ls.current = target.colname
  ################################################################################ 
  

    if (tmp.type=='neale'){
      gwas_sumstats = read_tsv(tmp.fname) %>% 
        left_join(.,ref.neale,by='variant') %>% 
        filter(chr!='X') %>%
        mutate(chr=as.numeric(chr)) %>% 
        select(SNP=rsid,Allele1=A2,Allele2=A1,
               Effect=beta,se,p=pval,Freq1=freq,CHR=chr,BP=pos) %>% 
        mutate(N=tmp.n) 
      
    }else if(tmp.type=='big40'){
      gwas_sumstats = read_delim(tmp.fname,delim=" ")%>% 
        mutate(pval=10^(-`pval(-log10)`),n=tmp.n) %>%
        left_join(.,ref.big40,by=c('chr'='chr','pos'='pos','rsid'='rsid','a1'='a1','a2'='a2')) %>% 
        filter(info>0.8,chr!='0X') %>% 
        mutate(chr=as.numeric(chr)) %>% 
        select(SNP=rsid,Allele1=a2,Allele2=a1,
               Effect=beta,se,p=pval,Freq1=af,CHR=chr,BP=pos) %>% 
        mutate(N=tmp.n) %>% 
        as.data.frame
    }else if(tmp.type=='mtag'){
      gwas_sumstats = read_tsv(tmp.fname)%>% 
        left_join(.,ref.big40,by=c('BP'='pos','SNP'='rsid','A1'='a1','A2'='a2')) %>% 
        select(SNP,Allele1=A1,Allele2=A2,
               Effect=mtag_beta,se=mtag_se,p=mtag_pval,Freq1=meta_freq,CHR,BP) %>%
        mutate(N=tmp.n) %>% 
        as.data.frame
    }else if(tmp.type=='local'){
      gwas_sumstats = read_tsv(tmp.fname) %>% 
        as.data.frame
    }else{
      print('No loading type')
    }
    cat('Sumstats loaded\n')

  
    cat('Processing as exposure\n')
    ### find top SNPs
    SigniSNP=gwas_sumstats %>% 
      mutate(NewAllele1=ifelse(Effect>=0,Allele1,Allele2),
             NewAllele2=ifelse(Effect>=0,Allele2,Allele1),
             NewFreq1=ifelse(Effect>=0,Freq1,1-Freq1),
             NewEffect=ifelse(Effect>=0,Effect,-1*Effect)) %>% 
      as.data.frame %>% 
      select(SNP,effect_allele=NewAllele1,other_allele=NewAllele2,
             eaf=NewFreq1,beta=NewEffect,se,pval=p,samplesize=N) %>% 
      mutate(units='unit',Phenotype=target.colname) %>% 
      .[.$SNP %in% ls.snp,]
    
    exposure_dat = format_data(SigniSNP, type="exposure")

  return(exposure_dat)
  gc()
}

extract_exposure_step2 <- function(tmp.targetadat,snp.step2){
  output = tmp.targetadat %>% 
    .[.$SNP %in% snp.step2,]
  return(output)
}


# Main process ------------------------------------------------------------

# Load exposure data
dat.tmp =
  inputs.gwas %>% split(.,seq(nrow(.))) %>% 
  pblapply(.,process_gwas,
           tmp_process.exposure=do.exposure,
           tmp_process.outcome=do.outcome,ls.snp=ls.snp_include)


# Clump again using the first exposure

tmp.exposure_dat <- clump_data(dat.tmp[[1]],clump_kb=250,clump_r2=0.001)
ls.snp_include.step2 <- tmp.exposure_dat$SNP

# Reformat and re-select SNPs for other exposure data
ls.exposure.input = data.frame(fname=paste0(output.path,'/',ls.exposure,'.tops'),
                               exposure.name=ls.exposure,
                               stringsAsFactors = F)
exposure.dat.stp2= dat.tmp %>% 
  lapply(.,extract_exposure_step2,snp.step2=ls.snp_include.step2) %>% 
  bind_rows

# Write exposure dat
write_tsv(exposure.dat.stp2,
          file=paste0(d.output,'/multivar.exposure_dat'),col_names=T)
write.table(ls.snp_include.step2,
            file = paste0(d.output,'/multivar.snps.exposure'),
            col.names = F,sep='\n',row.names = F,quote=F)

rm(list=ls())